// @okx-dex/okx-dex-sdk@1.0.17/dist/examples/solana/solana-instruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../../index");
const web3_js_1 = require("@solana/web3.js");
require("dotenv/config");
const wallet_1 = require("../../core/wallet");
async function main() {
    var _a, _b, _c, _d, _e;
    if (!process.env.SOLANA_RPC_URL)
        throw new Error('Missing SOLANA_RPC_URL');
    if (!process.env.SOLANA_PRIVATE_KEY)
        throw new Error('Missing SOLANA_PRIVATE_KEY');
    const connection = new web3_js_1.Connection(process.env.SOLANA_RPC_URL);
    const wallet = (0, wallet_1.createWallet)(process.env.SOLANA_PRIVATE_KEY, connection);
    const client = new index_1.OKXDexClient({
        apiKey: process.env.OKX_API_KEY,
        secretKey: process.env.OKX_SECRET_KEY,
        apiPassphrase: process.env.OKX_API_PASSPHRASE,
        projectId: process.env.OKX_PROJECT_ID,
        solana: { wallet },
    });
    const chainId = '501';
    const fromTokenAddress = '11111111111111111111111111111111'; // SOL
    const toTokenAddress = 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'; // BONK
    const resp = await client.dex.getSolanaSwapInstruction({
        chainId,
        fromTokenAddress,
        toTokenAddress,
        amount: '10000000', // 0.01 SOL
        userWalletAddress: wallet.publicKey.toString(),
        slippage: '0.1',
    });
    console.log('swap-instruction response :', resp);
    if (!resp.data || resp.data.length === 0) {
        console.error('swap-instruction empty');
        console.error(JSON.stringify(resp, null, 2));
        return;
    }
    const instr = resp.data[0];
    console.log('routerResult:', (_c = (_b = (_a = instr === null || instr === void 0 ? void 0 : instr.routerResult) === null || _a === void 0 ? void 0 : _a.dexRouterList) === null || _b === void 0 ? void 0 : _b[0]) === null || _c === void 0 ? void 0 : _c.router);
    console.log('instruction count:', (_e = (_d = instr === null || instr === void 0 ? void 0 : instr.instructionLists) === null || _d === void 0 ? void 0 : _d.length) !== null && _e !== void 0 ? _e : 0);
    const result = await client.dex.executeSolanaSwapInstructions({
        chainId,
        fromTokenAddress,
        toTokenAddress,
        amount: '10000000',
        userWalletAddress: wallet.publicKey.toString(),
        slippage: '0.01',
    });
    console.log('Tx Result:', result);
}
main().catch((e) => {
    console.error('Error:', e);
    process.exit(1);
});
