// @quantish/kalshi-sdk@1.0.5/dist/cli.js
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// src/cli.ts
var import_commander = require("commander");
var readline = __toESM(require("readline"));
var fs = __toESM(require("fs"));
var path = __toESM(require("path"));
var crypto = __toESM(require("crypto"));
var https = __toESM(require("https"));
var program = new import_commander.Command();
var KALSHI_MCP_URL = "https://kalshi-mcp-server-production.up.railway.app";
function makeRequest(method, path2, body, headers) {
  return new Promise((resolve, reject) => {
    const url = new URL(path2, KALSHI_MCP_URL);
    const options = {
      hostname: url.hostname,
      port: 443,
      path: url.pathname + url.search,
      method,
      headers: {
        "Content-Type": "application/json",
        ...headers
      }
    };
    const req = https.request(options, (res) => {
      let data = "";
      res.on("data", (chunk) => data += chunk);
      res.on("end", () => {
        try {
          resolve({ status: res.statusCode || 0, data: JSON.parse(data) });
        } catch {
          resolve({ status: res.statusCode || 0, data });
        }
      });
    });
    req.on("error", reject);
    req.setTimeout(3e4, () => {
      req.destroy();
      reject(new Error("Request timeout"));
    });
    if (body) req.write(JSON.stringify(body));
    req.end();
  });
}
async function mcpCall(toolName, args = {}, apiKey) {
  const headers = {};
  if (apiKey) headers["x-api-key"] = apiKey;
  return makeRequest("POST", "/mcp", {
    jsonrpc: "2.0",
    id: Date.now(),
    method: "tools/call",
    params: { name: toolName, arguments: args }
  }, headers);
}
function encryptWalletForImport(privateKeyBase58, password) {
  if (password.length < 12) {
    throw new Error("Password must be at least 12 characters");
  }
  const salt = crypto.randomBytes(32);
  const derivedKey = crypto.pbkdf2Sync(password, salt, 1e5, 32, "sha512");
  const iv = crypto.randomBytes(16);
  const cipher = crypto.createCipheriv("aes-256-gcm", derivedKey, iv);
  let encrypted = cipher.update(privateKeyBase58, "utf8", "hex");
  encrypted += cipher.final("hex");
  const authTag = cipher.getAuthTag().toString("hex");
  return {
    encryptedKey: `${encrypted}:${authTag}`,
    salt: salt.toString("hex"),
    iv: iv.toString("hex"),
    version: "1.0"
  };
}
program.name("quantish-kalshi").description("Quantish Kalshi SDK CLI - Prediction market trading on Solana").version("1.0.0");
program.command("setup").description("Interactive setup wizard for Quantish Kalshi SDK").action(async () => {
  console.log("\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551     QUANTISH KALSHI SDK - SETUP WIZARD                        \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  const question = (query) => new Promise((resolve) => rl.question(query, resolve));
  try {
    console.log("Step 1: Access Code");
    console.log("\u2500".repeat(50));
    const accessCode = await question("Enter your access code (KALSHI-XXXX-XXXX-XXXX): ");
    if (!accessCode.startsWith("KALSHI-")) {
      console.error("\u274C Invalid access code format");
      rl.close();
      return;
    }
    console.log("\nStep 2: User Identification");
    console.log("\u2500".repeat(50));
    const externalId = await question("Enter a unique identifier for your account: ");
    if (!externalId || externalId.length < 3) {
      console.error("\u274C External ID must be at least 3 characters");
      rl.close();
      return;
    }
    console.log("\n\u23F3 Requesting API credentials...");
    const result = await mcpCall("request_api_key", {
      accessCode,
      externalId,
      keyName: "CLI Setup"
    });
    const content = JSON.parse(result.data.result?.content?.[0]?.text || "{}");
    if (!content.apiKey) {
      console.error("\u274C Failed to get API key:", content.error || result.data.error?.message);
      rl.close();
      return;
    }
    console.log("\n\u2705 API credentials received!");
    console.log("\u2500".repeat(50));
    console.log("API Key:", content.apiKey);
    console.log("API Secret:", content.apiSecret);
    console.log("User ID:", content.userId);
    console.log("\nStep 3: Wallet Setup");
    console.log("\u2500".repeat(50));
    const walletChoice = await question("Create new wallet or import existing? (new/import): ");
    if (walletChoice.toLowerCase() === "import") {
      console.log("\n\u{1F4CB} To import your wallet:");
      console.log("   1. Export your private key from Phantom/Solflare");
      console.log("   2. Run: quantish-kalshi import-wallet");
      console.log("   3. Follow the prompts to securely encrypt and import");
    } else {
      console.log("\n\u23F3 Creating new Solana wallet...");
      const walletResult = await mcpCall("setup_wallet", {}, content.apiKey);
      const walletContent = JSON.parse(walletResult.data.result?.content?.[0]?.text || "{}");
      if (walletContent.wallet) {
        console.log("\n\u2705 Wallet created!");
        console.log("Public Key:", walletContent.wallet.publicKey);
        console.log("\n\u{1F4B0} Fund this address with SOL and USDC to start trading");
      } else {
        console.log("\u274C Failed to create wallet:", walletContent.message || walletContent.error);
      }
    }
    console.log("\n" + "\u2550".repeat(60));
    console.log("MCP CONFIGURATION FOR CURSOR IDE");
    console.log("\u2550".repeat(60));
    console.log("\nAdd this to your ~/.cursor/mcp.json:\n");
    console.log(JSON.stringify({
      quantish_kalshi: {
        url: KALSHI_MCP_URL + "/mcp",
        headers: {
          "x-api-key": content.apiKey
        }
      }
    }, null, 2));
    console.log("\n\u2705 Setup complete!");
  } catch (error) {
    console.error("\u274C Setup failed:", error.message);
  } finally {
    rl.close();
  }
});
program.command("import-wallet").description("Securely import an existing Solana wallet").option("-k, --api-key <key>", "Your Quantish Kalshi API key").action(async (options) => {
  console.log("\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551     SECURE WALLET IMPORT                                      \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  console.log("\u26A0\uFE0F  SECURITY NOTICE:");
  console.log("   \u2022 Your private key will be encrypted LOCALLY");
  console.log("   \u2022 Only the encrypted blob is sent to our servers");
  console.log("   \u2022 We CANNOT decrypt it without your password");
  console.log("   \u2022 You will need your password for every transaction\n");
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  const question = (query) => new Promise((resolve) => rl.question(query, resolve));
  try {
    let apiKey = options.apiKey;
    if (!apiKey) {
      apiKey = await question("Enter your API key (pk_kalshi_...): ");
    }
    if (!apiKey.startsWith("pk_kalshi_")) {
      console.error("\u274C Invalid API key format");
      rl.close();
      return;
    }
    console.log("\n\u{1F4DD} Enter your Solana private key (Base58 format):");
    console.log("   (This is typically 87-88 characters long)\n");
    const privateKey = await question("Private Key: ");
    if (privateKey.length < 80) {
      console.error("\u274C Invalid private key format");
      rl.close();
      return;
    }
    const publicKey = await question("Public Key (for verification): ");
    console.log("\n\u{1F510} Create a strong password (min 12 characters):");
    console.log("   This password encrypts your private key locally.");
    console.log("   You will need it for EVERY transaction.\n");
    const password = await question("Password: ");
    if (password.length < 12) {
      console.error("\u274C Password must be at least 12 characters");
      rl.close();
      return;
    }
    const confirmPassword = await question("Confirm Password: ");
    if (password !== confirmPassword) {
      console.error("\u274C Passwords do not match");
      rl.close();
      return;
    }
    console.log("\n\u23F3 Encrypting private key locally...");
    const bundle = encryptWalletForImport(privateKey, password);
    bundle.publicKey = publicKey;
    console.log("\u2705 Private key encrypted (never sent unencrypted)");
    console.log("\n\u23F3 Importing encrypted wallet...");
    const result = await mcpCall("import_wallet", bundle, apiKey);
    const content = JSON.parse(result.data.result?.content?.[0]?.text || "{}");
    if (content.wallet) {
      console.log("\n\u2705 Wallet imported successfully!");
      console.log("Public Key:", content.wallet.publicKey);
      console.log("Type:", content.wallet.type);
      console.log("\n\u26A0\uFE0F  Remember: You need your password for every transaction");
    } else {
      console.error("\u274C Import failed:", content.error || result.data.error?.message);
    }
  } catch (error) {
    console.error("\u274C Import failed:", error.message);
  } finally {
    rl.close();
  }
});
program.command("config").description("Generate MCP configuration for Cursor IDE").option("-k, --api-key <key>", "Your Quantish Kalshi API key").action(async (options) => {
  console.log("\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551     MCP CONFIGURATION GENERATOR                               \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  let apiKey = options.apiKey;
  if (!apiKey) {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
    apiKey = await new Promise(
      (resolve) => rl.question("Enter your API key (pk_kalshi_...): ", resolve)
    );
    rl.close();
  }
  if (!apiKey.startsWith("pk_kalshi_")) {
    console.error("\u274C Invalid API key format");
    return;
  }
  console.log("\n\u{1F4CB} CURSOR IDE CONFIGURATION");
  console.log("\u2550".repeat(50));
  console.log("\nAdd this to ~/.cursor/mcp.json:\n");
  const config = {
    quantish_kalshi: {
      url: KALSHI_MCP_URL + "/mcp",
      headers: {
        "x-api-key": apiKey
      }
    }
  };
  console.log(JSON.stringify(config, null, 2));
  console.log("\n\u{1F4CB} CLAUDE DESKTOP CONFIGURATION");
  console.log("\u2550".repeat(50));
  console.log("\nAdd this to claude_desktop_config.json:\n");
  const claudeConfig = {
    mcpServers: {
      quantish_kalshi: {
        command: "curl",
        args: [
          "-X",
          "POST",
          "-H",
          "Content-Type: application/json",
          "-H",
          `x-api-key: ${apiKey}`,
          KALSHI_MCP_URL + "/mcp"
        ]
      }
    }
  };
  console.log(JSON.stringify(claudeConfig, null, 2));
});
program.command("enable").description("Automatically add Kalshi SDK to Cursor MCP config").option("-k, --api-key <key>", "Your Quantish Kalshi API key").action(async (options) => {
  console.log("\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551     ONE-CLICK CURSOR INTEGRATION                              \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  let apiKey = options.apiKey;
  if (!apiKey) {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
    apiKey = await new Promise(
      (resolve) => rl.question("Enter your API key (pk_kalshi_...): ", resolve)
    );
    rl.close();
  }
  if (!apiKey.startsWith("pk_kalshi_")) {
    console.error("\u274C Invalid API key format");
    return;
  }
  const homeDir = process.env.HOME || process.env.USERPROFILE || "";
  const cursorConfigPath = path.join(homeDir, ".cursor", "mcp.json");
  try {
    let existingConfig = { mcpServers: {} };
    if (fs.existsSync(cursorConfigPath)) {
      const rawConfig = fs.readFileSync(cursorConfigPath, "utf-8");
      existingConfig = JSON.parse(rawConfig);
      if (!existingConfig.mcpServers) {
        existingConfig.mcpServers = {};
      }
    }
    const backupPath = `${cursorConfigPath}.backup.${Date.now()}`;
    if (fs.existsSync(cursorConfigPath)) {
      fs.copyFileSync(cursorConfigPath, backupPath);
      console.log("\u2705 Backed up existing config to:", backupPath);
    }
    existingConfig.mcpServers.quantish_kalshi = {
      url: KALSHI_MCP_URL + "/mcp",
      headers: {
        "x-api-key": apiKey
      }
    };
    const cursorDir = path.dirname(cursorConfigPath);
    if (!fs.existsSync(cursorDir)) {
      fs.mkdirSync(cursorDir, { recursive: true });
    }
    fs.writeFileSync(cursorConfigPath, JSON.stringify(existingConfig, null, 2));
    console.log("\u2705 Kalshi SDK added to Cursor MCP config!");
    console.log("   Config file:", cursorConfigPath);
    console.log("\n\u26A0\uFE0F  IMPORTANT: Restart Cursor IDE to apply changes");
  } catch (error) {
    console.error("\u274C Failed to update config:", error.message);
    console.log("\nManual configuration:");
    console.log(JSON.stringify({
      quantish_kalshi: {
        url: KALSHI_MCP_URL + "/mcp",
        headers: { "x-api-key": apiKey }
      }
    }, null, 2));
  }
});
program.command("test").description("Test your Kalshi SDK connection and wallet").option("-k, --api-key <key>", "Your Quantish Kalshi API key").action(async (options) => {
  console.log("\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551     KALSHI SDK CONNECTION TEST                                \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  let apiKey = options.apiKey;
  if (!apiKey) {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
    apiKey = await new Promise(
      (resolve) => rl.question("Enter your API key (pk_kalshi_...): ", resolve)
    );
    rl.close();
  }
  console.log("\u23F3 Testing connection...\n");
  console.log("1. Health Check...");
  try {
    const health = await makeRequest("GET", "/health");
    console.log("   Status:", health.status === 200 ? "\u2705 Healthy" : "\u274C Failed");
  } catch (e) {
    console.log("   \u274C Error:", e.message);
  }
  console.log("2. MCP Protocol...");
  try {
    const init = await makeRequest("POST", "/mcp", {
      jsonrpc: "2.0",
      id: 1,
      method: "initialize"
    });
    console.log("   Status:", init.data.result ? "\u2705 Working" : "\u274C Failed");
  } catch (e) {
    console.log("   \u274C Error:", e.message);
  }
  console.log("3. API Key Validation...");
  try {
    const result = await mcpCall("get_wallet_info", {}, apiKey);
    if (result.data.error?.message?.includes("Invalid")) {
      console.log("   \u274C Invalid API key");
    } else {
      console.log("   \u2705 API key is valid");
    }
  } catch (e) {
    console.log("   \u274C Error:", e.message);
  }
  console.log("4. Wallet Status...");
  try {
    const result = await mcpCall("get_wallet_info", {}, apiKey);
    const content = JSON.parse(result.data.result?.content?.[0]?.text || "{}");
    if (content.wallet) {
      console.log("   \u2705 Wallet found:", content.wallet.publicKey.substring(0, 20) + "...");
      console.log("   Type:", content.wallet.type);
    } else {
      console.log("   \u2139\uFE0F  No wallet setup yet");
    }
  } catch (e) {
    console.log("   \u274C Error:", e.message);
  }
  console.log("5. Balances...");
  try {
    const result = await mcpCall("get_balances", {}, apiKey);
    const content = JSON.parse(result.data.result?.content?.[0]?.text || "{}");
    if (content.balances) {
      console.log("   SOL:", content.balances.sol);
      console.log("   USDC:", content.balances.usdc);
    } else {
      console.log("   \u2139\uFE0F  No wallet to check balances");
    }
  } catch (e) {
    console.log("   \u274C Error:", e.message);
  }
  console.log("\n\u2705 Connection test complete!");
});
program.command("balance").description("Check your wallet balance").option("-k, --api-key <key>", "Your Quantish Kalshi API key").action(async (options) => {
  let apiKey = options.apiKey;
  if (!apiKey) {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
    apiKey = await new Promise(
      (resolve) => rl.question("Enter your API key: ", resolve)
    );
    rl.close();
  }
  try {
    const result = await mcpCall("get_balances", {}, apiKey);
    const content = JSON.parse(result.data.result?.content?.[0]?.text || "{}");
    if (content.balances) {
      console.log("\n\u{1F4B0} WALLET BALANCES");
      console.log("\u2550".repeat(30));
      console.log("Public Key:", content.publicKey);
      console.log("SOL:", content.balances.sol);
      console.log("USDC:", content.balances.usdc);
    } else {
      console.log("\u274C Could not fetch balances");
    }
  } catch (e) {
    console.error("\u274C Error:", e.message);
  }
});
program.parse(process.argv);
