// @quantish/kalshi-sdk@1.0.5/dist/index.mjs
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// src/index.ts
import express from "express";
import cors from "cors";
import helmet from "helmet";
import rateLimit from "express-rate-limit";

// src/config/index.ts
import dotenv from "dotenv";
dotenv.config();
var config = {
  server: {
    port: parseInt(process.env.PORT || "3002", 10),
    nodeEnv: process.env.NODE_ENV || "development"
  },
  database: {
    url: process.env.DATABASE_URL || ""
  },
  encryption: {
    key: process.env.ENCRYPTION_KEY || ""
  },
  solana: {
    rpcUrl: process.env.SOLANA_RPC_URL || "https://api.mainnet-beta.solana.com",
    rpcFallback: process.env.SOLANA_RPC_FALLBACK || "",
    commitment: "confirmed"
  },
  dflow: {
    // Production DFlow API endpoints for Kalshi trading on Solana
    apiUrl: process.env.DFLOW_API_URL || "https://quote-api.dflow.net",
    metadataApiUrl: process.env.DFLOW_METADATA_API_URL || "https://prediction-markets-api.dflow.net",
    apiKey: process.env.DFLOW_API_KEY || ""
  },
  admin: {
    apiKey: process.env.ADMIN_API_KEY || ""
  },
  rateLimit: {
    windowMs: parseInt(process.env.RATE_LIMIT_WINDOW_MS || "60000", 10),
    maxRequests: parseInt(process.env.RATE_LIMIT_MAX_REQUESTS || "60", 10)
  },
  // Known SPL token addresses on Solana
  tokens: {
    USDC: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
    // Native USDC on Solana
    USDT: "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB",
    SOL: "So11111111111111111111111111111111111111112"
    // Wrapped SOL
  }
};
function validateConfig() {
  const required = [
    { key: "DATABASE_URL", value: config.database.url },
    { key: "ENCRYPTION_KEY", value: config.encryption.key }
  ];
  const missing = required.filter(({ value }) => !value);
  if (missing.length > 0) {
    throw new Error(
      `Missing required environment variables: ${missing.map((m) => m.key).join(", ")}`
    );
  }
  if (config.encryption.key.length !== 64) {
    throw new Error("ENCRYPTION_KEY must be a 64-character hex string (32 bytes)");
  }
}

// src/routes/mcp.ts
import { Router } from "express";

// src/services/apikey.service.ts
import crypto2 from "crypto";

// src/db/index.ts
import { PrismaClient } from "@prisma/client";
var prisma = null;
function getPrismaClient() {
  if (!prisma) {
    prisma = new PrismaClient({
      log: process.env.NODE_ENV === "development" ? ["query", "error", "warn"] : ["error"]
    });
  }
  return prisma;
}
async function disconnectPrisma() {
  if (prisma) {
    await prisma.$disconnect();
    prisma = null;
  }
}

// src/crypto/encryption.ts
import crypto from "crypto";
var EncryptionService = class {
  key;
  algorithm = "aes-256-gcm";
  ivLength = 16;
  authTagLength = 16;
  constructor() {
    if (!config.encryption.key || config.encryption.key.length !== 64) {
      throw new Error("ENCRYPTION_KEY must be a 64-character hex string (32 bytes)");
    }
    this.key = Buffer.from(config.encryption.key, "hex");
  }
  /**
   * Encrypt plaintext using AES-256-GCM with random IV
   * Format: iv:authTag:ciphertext (all hex encoded)
   */
  encrypt(plaintext) {
    const iv = crypto.randomBytes(this.ivLength);
    const cipher = crypto.createCipheriv(this.algorithm, this.key, iv);
    let encrypted = cipher.update(plaintext, "utf8", "hex");
    encrypted += cipher.final("hex");
    const authTag = cipher.getAuthTag().toString("hex");
    return `${iv.toString("hex")}:${authTag}:${encrypted}`;
  }
  /**
   * Decrypt ciphertext encrypted with encrypt()
   */
  decrypt(ciphertext) {
    const parts = ciphertext.split(":");
    if (parts.length !== 3) {
      throw new Error("Invalid ciphertext format");
    }
    const [ivHex, authTagHex, encryptedHex] = parts;
    const iv = Buffer.from(ivHex, "hex");
    const authTag = Buffer.from(authTagHex, "hex");
    const decipher = crypto.createDecipheriv(this.algorithm, this.key, iv);
    decipher.setAuthTag(authTag);
    let decrypted = decipher.update(encryptedHex, "hex", "utf8");
    decrypted += decipher.final("utf8");
    return decrypted;
  }
  /**
   * Hash a value using SHA-256
   */
  hash(value) {
    return crypto.createHash("sha256").update(value).digest("hex");
  }
  /**
   * Generate a secure random string
   */
  generateSecureRandom(bytes = 32) {
    return crypto.randomBytes(bytes).toString("hex");
  }
};
var encryptionServiceInstance = null;
function getEncryptionService() {
  if (!encryptionServiceInstance) {
    encryptionServiceInstance = new EncryptionService();
  }
  return encryptionServiceInstance;
}

// src/services/apikey.service.ts
var ApiKeyService = class {
  /**
   * Generate a new API key for a user
   */
  async createApiKey(userId, name) {
    const prisma2 = getPrismaClient();
    const encryption = getEncryptionService();
    const keyRandom = crypto2.randomBytes(24).toString("base64url");
    const apiKey = `pk_kalshi_${keyRandom}`;
    const keyPrefix = apiKey.substring(0, 16);
    const apiSecret = `sk_kalshi_${crypto2.randomBytes(32).toString("base64url")}`;
    const keyHash = this.hashKey(apiKey);
    const secretHash = this.hashKey(apiSecret);
    const keyRecord = await prisma2.userApiKey.create({
      data: {
        userId,
        keyHash,
        keyPrefix,
        apiSecret: secretHash,
        name,
        isActive: true
      }
    });
    return {
      apiKey,
      apiSecret,
      keyPrefix,
      keyId: keyRecord.id
    };
  }
  /**
   * Validate an API key and return associated user
   */
  async validateApiKey(apiKey) {
    const prisma2 = getPrismaClient();
    if (!apiKey || !apiKey.startsWith("pk_kalshi_")) {
      return { isValid: false, message: "Invalid API key format" };
    }
    const keyHash = this.hashKey(apiKey);
    const keyRecord = await prisma2.userApiKey.findUnique({
      where: { keyHash },
      include: { user: true }
    });
    if (!keyRecord) {
      return { isValid: false, message: "API key not found" };
    }
    if (!keyRecord.isActive) {
      return { isValid: false, message: "API key is inactive" };
    }
    if (keyRecord.expiresAt && keyRecord.expiresAt < /* @__PURE__ */ new Date()) {
      return { isValid: false, message: "API key has expired" };
    }
    await prisma2.userApiKey.update({
      where: { id: keyRecord.id },
      data: { lastUsedAt: /* @__PURE__ */ new Date() }
    });
    return {
      isValid: true,
      userId: keyRecord.userId,
      keyRecord
    };
  }
  /**
   * Validate HMAC signature for request
   */
  validateHmacSignature(apiSecret, timestamp, method, path, body, providedSignature) {
    const requestTime = parseInt(timestamp, 10);
    const now = Date.now();
    if (Math.abs(now - requestTime) > 3e4) {
      return false;
    }
    const message = `${timestamp}${method}${path}${body}`;
    const expectedSignature = crypto2.createHmac("sha256", apiSecret).update(message).digest("base64");
    return crypto2.timingSafeEqual(
      Buffer.from(providedSignature),
      Buffer.from(expectedSignature)
    );
  }
  /**
   * Create HMAC signature (for client reference)
   */
  createHmacSignature(apiSecret, timestamp, method, path, body) {
    const message = `${timestamp}${method}${path}${body}`;
    return crypto2.createHmac("sha256", apiSecret).update(message).digest("base64");
  }
  /**
   * List all API keys for a user
   */
  async listApiKeys(userId) {
    const prisma2 = getPrismaClient();
    const keys = await prisma2.userApiKey.findMany({
      where: { userId },
      select: {
        id: true,
        keyPrefix: true,
        name: true,
        isActive: true,
        lastUsedAt: true,
        createdAt: true
      },
      orderBy: { createdAt: "desc" }
    });
    return keys;
  }
  /**
   * Revoke an API key
   */
  async revokeApiKey(userId, keyId) {
    const prisma2 = getPrismaClient();
    const key = await prisma2.userApiKey.findFirst({
      where: { id: keyId, userId }
    });
    if (!key) {
      return false;
    }
    await prisma2.userApiKey.update({
      where: { id: keyId },
      data: { isActive: false }
    });
    return true;
  }
  /**
   * Hash an API key or secret
   */
  hashKey(key) {
    return crypto2.createHash("sha256").update(key).digest("hex");
  }
};
var apiKeyServiceInstance = null;
function getApiKeyService() {
  if (!apiKeyServiceInstance) {
    apiKeyServiceInstance = new ApiKeyService();
  }
  return apiKeyServiceInstance;
}

// src/services/accesscode.service.ts
import { v4 as uuidv4 } from "uuid";
var AccessCodeService = class {
  /**
   * Create a new access code
   */
  async createAccessCode(data) {
    const prisma2 = getPrismaClient();
    const randomPart = uuidv4().toUpperCase().replace(/-/g, "").substring(0, 12).match(/.{1,4}/g)?.join("-");
    const code = `KALSHI-${randomPart}`;
    const expiresAt = data.expiresInDays ? new Date(Date.now() + data.expiresInDays * 24 * 60 * 60 * 1e3) : null;
    const accessCode = await prisma2.accessCode.create({
      data: {
        code,
        developerName: data.developerName,
        developerEmail: data.developerEmail,
        notes: data.notes,
        maxUses: data.maxUses || 1,
        expiresAt,
        createdBy: data.createdBy
      }
    });
    return accessCode;
  }
  /**
   * Validate and use an access code
   */
  async validateAndUseAccessCode(code, externalId) {
    const prisma2 = getPrismaClient();
    const accessCode = await prisma2.accessCode.findUnique({
      where: { code }
    });
    if (!accessCode) {
      return { isValid: false, message: "Invalid access code" };
    }
    if (!accessCode.isActive) {
      return { isValid: false, message: "Access code is inactive" };
    }
    if (accessCode.expiresAt && accessCode.expiresAt < /* @__PURE__ */ new Date()) {
      return { isValid: false, message: "Access code has expired" };
    }
    if (accessCode.maxUses !== -1 && accessCode.currentUses >= accessCode.maxUses) {
      return { isValid: false, message: "Access code has reached maximum uses" };
    }
    await prisma2.accessCode.update({
      where: { id: accessCode.id },
      data: { currentUses: { increment: 1 } }
    });
    return { isValid: true, message: "Access code valid" };
  }
  /**
   * List all access codes
   */
  async listAccessCodes() {
    const prisma2 = getPrismaClient();
    return prisma2.accessCode.findMany({
      orderBy: { createdAt: "desc" }
    });
  }
  /**
   * Revoke an access code
   */
  async revokeAccessCode(codeOrId) {
    const prisma2 = getPrismaClient();
    try {
      const accessCode = await prisma2.accessCode.findFirst({
        where: {
          OR: [{ id: codeOrId }, { code: codeOrId }]
        }
      });
      if (!accessCode) return false;
      await prisma2.accessCode.update({
        where: { id: accessCode.id },
        data: { isActive: false }
      });
      return true;
    } catch {
      return false;
    }
  }
};
var accessCodeServiceInstance = null;
function getAccessCodeService() {
  if (!accessCodeServiceInstance) {
    accessCodeServiceInstance = new AccessCodeService();
  }
  return accessCodeServiceInstance;
}

// src/wallet/solana-wallet.service.ts
import {
  Keypair as Keypair2,
  PublicKey,
  Connection,
  Transaction,
  VersionedTransaction,
  sendAndConfirmTransaction,
  LAMPORTS_PER_SOL,
  SystemProgram
} from "@solana/web3.js";
import {
  getAssociatedTokenAddress,
  getAccount,
  TOKEN_PROGRAM_ID,
  createTransferInstruction,
  createAssociatedTokenAccountInstruction
} from "@solana/spl-token";
import bs582 from "bs58";

// src/crypto/wallet-import.ts
import crypto3 from "crypto";
import { Keypair } from "@solana/web3.js";
import bs58 from "bs58";
function decryptImportedWallet(bundle, password) {
  const [encryptedHex, authTagHex] = bundle.encryptedKey.split(":");
  if (!encryptedHex || !authTagHex) {
    throw new Error("Invalid encrypted key format");
  }
  const salt = Buffer.from(bundle.salt, "hex");
  const iv = Buffer.from(bundle.iv, "hex");
  const authTag = Buffer.from(authTagHex, "hex");
  const derivedKey = crypto3.pbkdf2Sync(
    password,
    salt,
    1e5,
    32,
    "sha512"
  );
  const decipher = crypto3.createDecipheriv("aes-256-gcm", derivedKey, iv);
  decipher.setAuthTag(authTag);
  let decrypted;
  try {
    decrypted = decipher.update(encryptedHex, "hex", "utf8");
    decrypted += decipher.final("utf8");
  } catch (error) {
    throw new Error("Decryption failed - incorrect password or corrupted data");
  }
  let keypair;
  try {
    const privateKeyBytes = bs58.decode(decrypted);
    keypair = Keypair.fromSecretKey(privateKeyBytes);
  } catch (error) {
    throw new Error("Decrypted key is invalid");
  }
  if (keypair.publicKey.toBase58() !== bundle.publicKey) {
    throw new Error("Public key mismatch - data may be corrupted");
  }
  return {
    keypair,
    publicKey: keypair.publicKey.toBase58()
  };
}
function verifyWalletImportBundle(bundle) {
  try {
    if (bundle.version !== "1.0") {
      return false;
    }
    if (!bundle.encryptedKey || !bundle.salt || !bundle.iv || !bundle.publicKey) {
      return false;
    }
    const publicKeyBytes = bs58.decode(bundle.publicKey);
    if (publicKeyBytes.length !== 32) {
      return false;
    }
    Buffer.from(bundle.salt, "hex");
    Buffer.from(bundle.iv, "hex");
    const [encryptedHex, authTagHex] = bundle.encryptedKey.split(":");
    if (!encryptedHex || !authTagHex) {
      return false;
    }
    Buffer.from(encryptedHex, "hex");
    Buffer.from(authTagHex, "hex");
    return true;
  } catch {
    return false;
  }
}
function getWalletExportInstructions() {
  return `
## How to Export Your Solana Wallet for Secure Import

### From Phantom Wallet:
1. Open Phantom wallet
2. Click the settings icon (gear) in the bottom right
3. Select your wallet
4. Click "Show Secret Recovery Phrase" or "Export Private Key"
5. Enter your password
6. Copy the private key (it's a long string of letters and numbers)

### From Solflare Wallet:
1. Open Solflare
2. Go to Settings \u2192 Security
3. Click "Export Private Key"
4. Authenticate with your password
5. Copy the private key

### Security Notes:
- NEVER share your raw private key with anyone
- Always encrypt it with a strong password before importing
- Use a unique password (at least 12 characters)
- Your encrypted key is safe to transmit - we cannot decrypt it without your password
- You'll need to provide your password each time you want to trade

### Encrypt and Import:
\`\`\`javascript
const { encryptWalletForImport } = require('@quantish/kalshi-sdk');

// On YOUR local machine (not on our servers)
const bundle = encryptWalletForImport(
  'your-private-key-here',
  'your-secure-password-12chars-minimum'
);

// Send this bundle to our API - it's safe, we can't decrypt it
console.log(bundle);
\`\`\`
`;
}

// src/wallet/solana-wallet.service.ts
var SolanaWalletService = class {
  connection;
  fallbackConnection = null;
  constructor() {
    this.connection = new Connection(config.solana.rpcUrl, config.solana.commitment);
    if (config.solana.rpcFallback) {
      this.fallbackConnection = new Connection(config.solana.rpcFallback, config.solana.commitment);
    }
  }
  /**
   * Generate a new Solana wallet for a user
   */
  async generateWallet(userId) {
    const prisma2 = getPrismaClient();
    const encryption = getEncryptionService();
    const keypair = Keypair2.generate();
    const publicKey = keypair.publicKey.toBase58();
    const privateKeyBase58 = bs582.encode(keypair.secretKey);
    const encryptedPrivateKey = encryption.encrypt(privateKeyBase58);
    await prisma2.user.update({
      where: { id: userId },
      data: {
        solanaPublicKey: publicKey,
        encryptedPrivateKey
      }
    });
    return {
      publicKey,
      type: "generated",
      createdAt: /* @__PURE__ */ new Date()
    };
  }
  /**
   * Import an existing wallet using encrypted bundle
   */
  async importWallet(userId, bundle) {
    const prisma2 = getPrismaClient();
    if (!verifyWalletImportBundle(bundle)) {
      throw new Error("Invalid wallet import bundle");
    }
    const existingUser = await prisma2.user.findFirst({
      where: {
        OR: [
          { solanaPublicKey: bundle.publicKey },
          { importedWalletPublicKey: bundle.publicKey }
        ]
      }
    });
    if (existingUser && existingUser.id !== userId) {
      throw new Error("This wallet is already registered to another user");
    }
    await prisma2.user.update({
      where: { id: userId },
      data: {
        importedWalletPublicKey: bundle.publicKey,
        importedWalletEncrypted: bundle.encryptedKey,
        importedWalletSalt: `${bundle.salt}:${bundle.iv}:${bundle.version}`,
        walletImportedAt: /* @__PURE__ */ new Date()
      }
    });
    return {
      publicKey: bundle.publicKey,
      type: "imported",
      createdAt: /* @__PURE__ */ new Date()
    };
  }
  /**
   * Get wallet info for a user
   */
  async getWalletInfo(userId) {
    const prisma2 = getPrismaClient();
    const user = await prisma2.user.findUnique({
      where: { id: userId },
      select: {
        solanaPublicKey: true,
        importedWalletPublicKey: true,
        walletImportedAt: true,
        createdAt: true
      }
    });
    if (!user) {
      return null;
    }
    if (user.importedWalletPublicKey) {
      return {
        publicKey: user.importedWalletPublicKey,
        type: "imported",
        createdAt: user.walletImportedAt || user.createdAt
      };
    }
    if (user.solanaPublicKey) {
      return {
        publicKey: user.solanaPublicKey,
        type: "generated",
        createdAt: user.createdAt
      };
    }
    return null;
  }
  /**
   * Get wallet balances
   */
  async getBalances(publicKey) {
    const pubkey = new PublicKey(publicKey);
    const solBalance = await this.connection.getBalance(pubkey);
    let usdcBalance = 0;
    try {
      const usdcMint = new PublicKey(config.tokens.USDC);
      const usdcAta = await getAssociatedTokenAddress(usdcMint, pubkey);
      const usdcAccount = await getAccount(this.connection, usdcAta);
      usdcBalance = Number(usdcAccount.amount) / 1e6;
    } catch {
    }
    return {
      sol: solBalance / LAMPORTS_PER_SOL,
      usdc: usdcBalance,
      tokens: {}
    };
  }
  /**
   * Get keypair for signing (for generated wallets)
   */
  async getKeypairForSigning(userId) {
    const prisma2 = getPrismaClient();
    const encryption = getEncryptionService();
    const user = await prisma2.user.findUnique({
      where: { id: userId },
      select: { encryptedPrivateKey: true }
    });
    if (!user?.encryptedPrivateKey) {
      throw new Error("No generated wallet found for this user. Use importedWallet signing for imported wallets.");
    }
    const privateKeyBase58 = encryption.decrypt(user.encryptedPrivateKey);
    const privateKeyBytes = bs582.decode(privateKeyBase58);
    return Keypair2.fromSecretKey(privateKeyBytes);
  }
  /**
   * Get keypair for signing (for imported wallets - requires password)
   */
  async getImportedWalletKeypair(userId, password) {
    const prisma2 = getPrismaClient();
    const user = await prisma2.user.findUnique({
      where: { id: userId },
      select: {
        importedWalletPublicKey: true,
        importedWalletEncrypted: true,
        importedWalletSalt: true
      }
    });
    if (!user?.importedWalletEncrypted || !user?.importedWalletSalt) {
      throw new Error("No imported wallet found for this user");
    }
    const [salt, iv, version] = user.importedWalletSalt.split(":");
    const bundle = {
      encryptedKey: user.importedWalletEncrypted,
      salt,
      iv,
      publicKey: user.importedWalletPublicKey,
      version: version || "1.0"
    };
    const { keypair } = decryptImportedWallet(bundle, password);
    return keypair;
  }
  /**
   * Sign and send a transaction
   */
  async signAndSendTransaction(userId, transaction, password) {
    const walletInfo = await this.getWalletInfo(userId);
    if (!walletInfo) {
      throw new Error("No wallet found for user");
    }
    let keypair;
    if (walletInfo.type === "imported") {
      if (!password) {
        throw new Error("Password required for imported wallet transactions");
      }
      keypair = await this.getImportedWalletKeypair(userId, password);
    } else {
      keypair = await this.getKeypairForSigning(userId);
    }
    try {
      if (transaction instanceof VersionedTransaction) {
        transaction.sign([keypair]);
        const signature = await this.connection.sendTransaction(transaction);
        await this.connection.confirmTransaction(signature, "confirmed");
        return signature;
      } else {
        const signature = await sendAndConfirmTransaction(
          this.connection,
          transaction,
          [keypair],
          { commitment: "confirmed" }
        );
        return signature;
      }
    } finally {
      keypair.secretKey.fill(0);
    }
  }
  /**
   * Get SPL token balance for a specific mint
   */
  async getTokenBalance(publicKey, mintAddress) {
    try {
      const pubkey = new PublicKey(publicKey);
      const mint = new PublicKey(mintAddress);
      const ata = await getAssociatedTokenAddress(mint, pubkey);
      const account = await getAccount(this.connection, ata);
      return Number(account.amount);
    } catch {
      return 0;
    }
  }
  /**
   * Get all SPL token accounts for a wallet
   */
  async getAllTokenAccounts(publicKey) {
    const pubkey = new PublicKey(publicKey);
    const tokenAccounts = await this.connection.getParsedTokenAccountsByOwner(
      pubkey,
      { programId: TOKEN_PROGRAM_ID }
    );
    return tokenAccounts.value.map((account) => {
      const parsed = account.account.data.parsed.info;
      return {
        mint: parsed.mint,
        balance: parsed.tokenAmount.amount,
        decimals: parsed.tokenAmount.decimals
      };
    });
  }
  /**
   * Get connection for external use
   */
  getConnection() {
    return this.connection;
  }
  /**
   * Send SOL to another wallet
   */
  async sendSol(userId, toAddress, solAmount, password) {
    const walletInfo = await this.getWalletInfo(userId);
    if (!walletInfo) {
      throw new Error("No wallet found for user");
    }
    if (walletInfo.type === "imported" && !password) {
      throw new Error("Password required for imported wallet transactions");
    }
    let toPubkey;
    try {
      toPubkey = new PublicKey(toAddress);
    } catch {
      throw new Error("Invalid Solana address");
    }
    const fromPubkey = new PublicKey(walletInfo.publicKey);
    const lamports = Math.floor(solAmount * LAMPORTS_PER_SOL);
    const balance = await this.connection.getBalance(fromPubkey);
    if (balance < lamports + 5e3) {
      throw new Error(`Insufficient SOL balance. Have: ${balance / LAMPORTS_PER_SOL}, need: ${solAmount} + fee`);
    }
    const transaction = new Transaction().add(
      SystemProgram.transfer({
        fromPubkey,
        toPubkey,
        lamports
      })
    );
    const { blockhash } = await this.connection.getLatestBlockhash();
    transaction.recentBlockhash = blockhash;
    transaction.feePayer = fromPubkey;
    const txSignature = await this.signAndSendTransaction(userId, transaction, password);
    return {
      txSignature,
      fromAddress: walletInfo.publicKey,
      toAddress,
      amount: solAmount,
      explorerUrl: `https://solscan.io/tx/${txSignature}`
    };
  }
  /**
   * Send SPL token to another wallet
   */
  async sendToken(userId, toAddress, mintAddress, amount, decimals, password) {
    const walletInfo = await this.getWalletInfo(userId);
    if (!walletInfo) {
      throw new Error("No wallet found for user");
    }
    if (walletInfo.type === "imported" && !password) {
      throw new Error("Password required for imported wallet transactions");
    }
    let toPubkey;
    let mintPubkey;
    try {
      toPubkey = new PublicKey(toAddress);
      mintPubkey = new PublicKey(mintAddress);
    } catch {
      throw new Error("Invalid Solana address");
    }
    const fromPubkey = new PublicKey(walletInfo.publicKey);
    const rawAmount = Math.floor(amount * Math.pow(10, decimals));
    const sourceAta = await getAssociatedTokenAddress(mintPubkey, fromPubkey);
    try {
      const sourceAccount = await getAccount(this.connection, sourceAta);
      if (Number(sourceAccount.amount) < rawAmount) {
        throw new Error(`Insufficient token balance. Have: ${Number(sourceAccount.amount) / Math.pow(10, decimals)}, need: ${amount}`);
      }
    } catch (e) {
      if (e.message?.includes("Insufficient")) {
        throw e;
      }
      throw new Error("Source token account not found");
    }
    const destAta = await getAssociatedTokenAddress(mintPubkey, toPubkey);
    const transaction = new Transaction();
    try {
      await getAccount(this.connection, destAta);
    } catch {
      transaction.add(
        createAssociatedTokenAccountInstruction(
          fromPubkey,
          destAta,
          toPubkey,
          mintPubkey
        )
      );
    }
    transaction.add(
      createTransferInstruction(
        sourceAta,
        destAta,
        fromPubkey,
        rawAmount
      )
    );
    const { blockhash } = await this.connection.getLatestBlockhash();
    transaction.recentBlockhash = blockhash;
    transaction.feePayer = fromPubkey;
    const txSignature = await this.signAndSendTransaction(userId, transaction, password);
    return {
      txSignature,
      fromAddress: walletInfo.publicKey,
      toAddress,
      mint: mintAddress,
      amount,
      explorerUrl: `https://solscan.io/tx/${txSignature}`
    };
  }
  /**
   * Helper: Send USDC
   */
  async sendUsdc(userId, toAddress, usdcAmount, password) {
    return this.sendToken(
      userId,
      toAddress,
      config.tokens.USDC,
      usdcAmount,
      6,
      // USDC decimals
      password
    );
  }
};
var walletServiceInstance = null;
function getSolanaWalletService() {
  if (!walletServiceInstance) {
    walletServiceInstance = new SolanaWalletService();
  }
  return walletServiceInstance;
}

// src/services/dflow-market.service.ts
import axios from "axios";
var DFlowMarketService = class {
  client;
  constructor() {
    const headers = {
      "Content-Type": "application/json"
    };
    if (config.dflow.apiKey) {
      headers["x-api-key"] = config.dflow.apiKey;
    }
    this.client = axios.create({
      baseURL: config.dflow.metadataApiUrl,
      timeout: 3e4,
      headers
    });
  }
  // ============================================
  // EVENTS API
  // ============================================
  /**
   * Get a single event by ticker
   */
  async getEvent(ticker) {
    const response = await this.client.get(`/api/v1/events/${ticker}`);
    return response.data;
  }
  /**
   * Get multiple events with optional filters
   * Note: Status filtering happens at market level, not event level
   */
  async getEvents(params) {
    const { marketStatus, limit, ...apiParams } = params || {};
    const requestedLimit = limit || 50;
    const fetchLimit = marketStatus && marketStatus !== "all" ? Math.max(200, requestedLimit * 4) : requestedLimit;
    const queryParams = {
      ...apiParams,
      limit: fetchLimit,
      withNestedMarkets: params?.withNestedMarkets ?? true
    };
    const response = await this.client.get("/api/v1/events", { params: queryParams });
    let events = response.data.events || [];
    if (marketStatus && marketStatus !== "all") {
      events = events.filter(
        (event) => event.markets?.some((market) => market.status === marketStatus)
      );
    }
    events = events.slice(0, requestedLimit);
    return { events, cursor: response.data.cursor };
  }
  /**
   * Search events by title or ticker
   * Fetches all events and filters client-side for best results
   * Includes all market statuses (active, inactive, finalized) for discovery
   */
  async searchEvents(query, limit = 20, marketStatus) {
    try {
      const response = await this.client.get("/api/v1/search/events", {
        params: { q: query, limit, withNestedMarkets: true }
      });
      let events = response.data.events || [];
      if (marketStatus && marketStatus !== "all") {
        events = events.filter(
          (e) => e.markets?.some((m) => m.status === marketStatus)
        );
      }
      return events;
    } catch (error) {
      if (error.response?.status === 404 || error.response?.status === 400) {
        const response = await this.client.get("/api/v1/events", {
          params: { withNestedMarkets: true, limit: 200 }
        });
        let events = response.data.events || [];
        const queryLower = query.toLowerCase();
        events = events.filter((e) => {
          const eventMatch = e.title?.toLowerCase().includes(queryLower) || e.ticker?.toLowerCase().includes(queryLower) || e.subtitle?.toLowerCase().includes(queryLower) || e.seriesTicker?.toLowerCase().includes(queryLower);
          const marketMatch = e.markets?.some(
            (m) => m.title?.toLowerCase().includes(queryLower) || m.ticker?.toLowerCase().includes(queryLower)
          );
          return eventMatch || marketMatch;
        });
        if (marketStatus && marketStatus !== "all") {
          events = events.filter(
            (e) => e.markets?.some((m) => m.status === marketStatus)
          );
        }
        return events.slice(0, limit);
      }
      throw error;
    }
  }
  /**
   * Get event forecast percentile history
   */
  async getEventForecastHistory(ticker) {
    const response = await this.client.get(`/api/v1/events/${ticker}/forecast-history`);
    return response.data;
  }
  /**
   * Get event candlestick data
   */
  async getEventCandlesticks(ticker, params) {
    const response = await this.client.get(`/api/v1/events/${ticker}/candlesticks`, { params });
    return response.data.candlesticks || [];
  }
  // ============================================
  // MARKETS API
  // ============================================
  /**
   * Get a single market by ticker
   * Note: DFlow API /api/v1/markets/{ticker} returns 404, so we use the list endpoint with filters
   */
  async getMarket(ticker) {
    const parts = ticker.split("-");
    const eventTicker = parts.length >= 3 ? parts.slice(0, -1).join("-") : ticker;
    const response = await this.client.get("/api/v1/markets", {
      params: { eventTicker, limit: 100 }
    });
    const markets = response.data.markets || [];
    const market = markets.find((m) => m.ticker === ticker);
    if (!market) {
      const eventsResponse = await this.client.get("/api/v1/events", {
        params: { withNestedMarkets: true, limit: 100 }
      });
      for (const event of eventsResponse.data.events || []) {
        const found = event.markets?.find((m) => m.ticker === ticker);
        if (found) {
          return found;
        }
      }
      return null;
    }
    return market;
  }
  /**
   * Get market by outcome mint address
   */
  async getMarketByMint(mintAddress) {
    const response = await this.client.get(`/api/v1/markets/by-mint/${mintAddress}`);
    return response.data;
  }
  /**
   * Get multiple markets
   */
  async getMarkets(params) {
    const response = await this.client.get("/api/v1/markets", { params });
    return response.data;
  }
  /**
   * Get outcome mints for a market
   */
  async getOutcomeMints(marketTicker) {
    const response = await this.client.get(`/api/v1/markets/${marketTicker}/mints`);
    return response.data;
  }
  /**
   * Filter outcome mints by criteria
   */
  async filterOutcomeMints(mints) {
    const response = await this.client.post("/api/v1/markets/filter-mints", { mints });
    return response.data.results || [];
  }
  /**
   * Get market candlestick data
   */
  async getMarketCandlesticks(ticker, params) {
    const response = await this.client.get(`/api/v1/markets/${ticker}/candlesticks`, { params });
    return response.data.candlesticks || [];
  }
  // ============================================
  // LIVE DATA API
  // Note: DFlow /api/v1/live/* endpoints return 404
  // We construct live data from market data instead
  // ============================================
  /**
   * Get live data for a market
   * Since DFlow live endpoints return 404, we get data from market details
   */
  async getLiveData(marketTicker) {
    const market = await this.getMarket(marketTicker);
    if (!market) {
      return null;
    }
    return {
      ticker: market.ticker,
      yesPrice: market.yesBid || market.yesAsk || 0,
      noPrice: market.noBid || market.noAsk || 0,
      volume24h: market.volume || 0,
      openInterest: market.openInterest || 0,
      yesBid: market.yesBid,
      yesAsk: market.yesAsk,
      noBid: market.noBid,
      noAsk: market.noAsk,
      status: market.status
    };
  }
  /**
   * Get live data for an event (all markets)
   */
  async getLiveDataByEvent(eventTicker) {
    const response = await this.client.get("/api/v1/markets", {
      params: { eventTicker, limit: 100 }
    });
    const markets = response.data.markets || [];
    return markets.map((market) => ({
      ticker: market.ticker,
      yesPrice: market.yesBid || market.yesAsk || 0,
      noPrice: market.noBid || market.noAsk || 0,
      volume24h: market.volume || 0,
      openInterest: market.openInterest || 0,
      yesBid: market.yesBid,
      yesAsk: market.yesAsk,
      noBid: market.noBid,
      noAsk: market.noAsk,
      status: market.status
    }));
  }
  /**
   * Get live data by mint address
   * Since DFlow live endpoints return 404, we search through events
   */
  async getLiveDataByMint(mintAddress) {
    const response = await this.client.get("/api/v1/events", {
      params: { withNestedMarkets: true, limit: 100 }
    });
    for (const event of response.data.events || []) {
      for (const market of event.markets || []) {
        if (market.accounts) {
          for (const settlement of Object.values(market.accounts)) {
            if (settlement.yesMint === mintAddress || settlement.noMint === mintAddress) {
              return {
                ticker: market.ticker,
                yesPrice: market.yesBid || market.yesAsk || 0,
                noPrice: market.noBid || market.noAsk || 0,
                volume24h: market.volume || 0,
                openInterest: market.openInterest || 0,
                yesBid: market.yesBid,
                yesAsk: market.yesAsk,
                noBid: market.noBid,
                noAsk: market.noAsk,
                status: market.status
              };
            }
          }
        }
      }
    }
    return null;
  }
  // ============================================
  // TRADES API
  // ============================================
  /**
   * Get recent trades for a market
   */
  async getTrades(marketTicker, params) {
    const response = await this.client.get(`/api/v1/trades/${marketTicker}`, { params });
    return response.data.trades || [];
  }
  /**
   * Get trades by mint
   */
  async getTradesByMint(mintAddress, params) {
    const response = await this.client.get(`/api/v1/trades/by-mint/${mintAddress}`, { params });
    return response.data.trades || [];
  }
  // ============================================
  // SERIES API
  // ============================================
  /**
   * Get all series templates
   */
  async getSeries() {
    const response = await this.client.get("/api/v1/series");
    return response.data.series || [];
  }
  /**
   * Get series by ticker
   */
  async getSeriesByTicker(ticker) {
    const response = await this.client.get(`/api/v1/series/${ticker}`);
    return response.data;
  }
  // ============================================
  // TAGS & CATEGORIES API
  // ============================================
  /**
   * Get tags organized by categories
   */
  async getTagsByCategories() {
    const response = await this.client.get("/api/v1/tags/by-categories");
    return response.data;
  }
  /**
   * Get sports filters
   */
  async getSportsFilters() {
    const response = await this.client.get("/api/v1/sports/filters");
    return response.data;
  }
  // ============================================
  // MARKET INITIALIZATION & REDEMPTION STATUS
  // ============================================
  /**
   * Check market initialization status
   * Returns whether the market is initialized (tokenized on-chain) and 
   * the outcome mints for trading.
   */
  async checkMarketInitialization(ticker, settlementMint = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v") {
    const market = await this.getMarket(ticker);
    if (!market) {
      return {
        isInitialized: false,
        marketTicker: ticker,
        status: "NOT_FOUND",
        note: "Market not found in DFlow. Check the ticker."
      };
    }
    const accountInfo = market.accounts?.[settlementMint];
    if (!accountInfo) {
      return {
        isInitialized: false,
        marketTicker: ticker,
        status: "NO_SETTLEMENT_MINT",
        note: `No account info for settlement mint ${settlementMint}. Market may not support this settlement token.`
      };
    }
    return {
      isInitialized: accountInfo.isInitialized ?? false,
      marketTicker: ticker,
      yesMint: accountInfo.yesMint,
      noMint: accountInfo.noMint,
      marketLedger: accountInfo.marketLedger,
      status: accountInfo.isInitialized ? "INITIALIZED" : "NOT_INITIALIZED",
      note: accountInfo.isInitialized ? "Market is tokenized and ready for trading." : "Market needs initialization. First trade or explicit initialization will create the outcome tokens on-chain (requires SOL fee)."
    };
  }
  /**
   * Check redemption status for a market
   * Determines if a market is settled and positions can be redeemed for stablecoins.
   */
  async checkRedemptionStatus(ticker, settlementMint = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v") {
    const market = await this.getMarket(ticker);
    if (!market) {
      return {
        isRedeemable: false,
        marketTicker: ticker,
        marketStatus: "NOT_FOUND",
        note: "Market not found."
      };
    }
    const accountInfo = market.accounts?.[settlementMint];
    if (!accountInfo) {
      return {
        isRedeemable: false,
        marketTicker: ticker,
        marketStatus: market.status,
        note: "No settlement account found for this market."
      };
    }
    const marketStatus = market.status;
    const result = market.result;
    const redemptionStatus = accountInfo.redemptionStatus;
    const isSettled = marketStatus === "finalized" || marketStatus === "determined";
    const isRedemptionOpen = redemptionStatus === "open";
    const isRedeemable = isSettled && isRedemptionOpen;
    let winningMint;
    let losingMint;
    if (result === "yes") {
      winningMint = accountInfo.yesMint;
      losingMint = accountInfo.noMint;
    } else if (result === "no") {
      winningMint = accountInfo.noMint;
      losingMint = accountInfo.yesMint;
    }
    let note;
    if (!isSettled) {
      note = `Market is ${marketStatus}. Wait for settlement before redemption.`;
    } else if (!isRedemptionOpen) {
      note = `Market is settled but redemption status is "${redemptionStatus}". Vault may still be funding.`;
    } else if (isRedeemable) {
      note = result ? `Redemption is open. ${result.toUpperCase()} tokens can be redeemed for full value.` : "Scalar outcome - both YES and NO tokens may be partially redeemable.";
    } else {
      note = "Cannot redeem at this time.";
    }
    return {
      isRedeemable,
      marketTicker: ticker,
      marketStatus,
      result,
      redemptionStatus,
      winningMint,
      losingMint,
      note
    };
  }
};
var marketServiceInstance = null;
function getDFlowMarketService() {
  if (!marketServiceInstance) {
    marketServiceInstance = new DFlowMarketService();
  }
  return marketServiceInstance;
}

// src/services/dflow-trade.service.ts
import axios2 from "axios";
import {
  Connection as Connection2,
  Transaction as Transaction2,
  VersionedTransaction as VersionedTransaction2
} from "@solana/web3.js";
var DFlowTradeService = class {
  client;
  connection;
  lastRequestTime = 0;
  minRequestInterval = 500;
  // 500ms between requests to avoid 429
  constructor() {
    const headers = {
      "Content-Type": "application/json"
    };
    if (config.dflow.apiKey) {
      headers["x-api-key"] = config.dflow.apiKey;
    }
    this.client = axios2.create({
      baseURL: config.dflow.apiUrl,
      timeout: 3e4,
      headers
    });
    this.client.interceptors.response.use(
      (response) => response,
      async (error) => {
        if (error.response?.status === 429) {
          const retryAfter = error.response.headers["retry-after"];
          const delay = retryAfter ? parseInt(retryAfter) * 1e3 : 2e3;
          console.log(`Rate limited, waiting ${delay}ms before retry...`);
          await new Promise((resolve) => setTimeout(resolve, delay));
          return this.client.request(error.config);
        }
        throw error;
      }
    );
    this.connection = new Connection2(config.solana.rpcUrl, config.solana.commitment);
  }
  async throttle() {
    const now = Date.now();
    const timeSinceLastRequest = now - this.lastRequestTime;
    if (timeSinceLastRequest < this.minRequestInterval) {
      await new Promise((resolve) => setTimeout(resolve, this.minRequestInterval - timeSinceLastRequest));
    }
    this.lastRequestTime = Date.now();
  }
  /**
   * Get a quote for a trade (includes transaction if user public key provided)
   */
  async getQuote(params) {
    await this.throttle();
    const queryParams = {
      inputMint: params.inputMint,
      outputMint: params.outputMint,
      amount: params.amount,
      slippageBps: params.slippageBps || 100
      // Default 1%
    };
    if (params.userPublicKey) {
      queryParams.userPublicKey = params.userPublicKey;
    }
    if (params.platformFeeBps) {
      queryParams.platformFeeBps = params.platformFeeBps;
    }
    if (params.feeAccount) {
      queryParams.feeAccount = params.feeAccount;
    }
    if (params.destinationWallet) {
      queryParams.destinationWallet = params.destinationWallet;
    }
    if (params.wrapAndUnwrapSol !== void 0) {
      queryParams.wrapAndUnwrapSol = params.wrapAndUnwrapSol;
    }
    if (params.onlyDirectRoutes !== void 0) {
      queryParams.onlyDirectRoutes = params.onlyDirectRoutes;
    }
    if (params.maxRouteLength !== void 0) {
      queryParams.maxRouteLength = params.maxRouteLength;
    }
    const response = await this.client.get("/order", { params: queryParams });
    return response.data;
  }
  /**
   * Get order status
   */
  async getOrderStatus(orderId) {
    const response = await this.client.get(`/order/${orderId}/status`);
    return response.data;
  }
  /**
   * Execute a trade for a user
   * 
   * For generated wallets: automatically signs and submits
   * For imported wallets: requires password parameter
   */
  async executeTrade(userId, params, password) {
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(userId);
    if (!walletInfo) {
      throw new Error("No wallet found for user");
    }
    if (walletInfo.type === "imported" && !password) {
      throw new Error("Password required for imported wallet transactions");
    }
    const quote = await this.getQuote({
      ...params,
      userPublicKey: walletInfo.publicKey
    });
    if (!quote.transaction) {
      throw new Error("No transaction returned from quote - insufficient liquidity or invalid parameters");
    }
    const txBuffer = Buffer.from(quote.transaction, "base64");
    let transaction;
    try {
      transaction = VersionedTransaction2.deserialize(txBuffer);
    } catch {
      transaction = Transaction2.from(txBuffer);
    }
    const txSignature = await walletService.signAndSendTransaction(
      userId,
      transaction,
      password
    );
    return {
      quote,
      txSignature,
      status: quote.executionMode === "sync" ? "success" : "pending"
    };
  }
  /**
   * Buy YES outcome tokens for a market
   */
  async buyYes(userId, marketTicker, yesOutcomeMint, usdcAmount, slippageBps, password) {
    return this.executeTrade(
      userId,
      {
        inputMint: config.tokens.USDC,
        outputMint: yesOutcomeMint,
        amount: usdcAmount * 1e6,
        // USDC has 6 decimals
        slippageBps
      },
      password
    );
  }
  /**
   * Buy NO outcome tokens for a market
   */
  async buyNo(userId, marketTicker, noOutcomeMint, usdcAmount, slippageBps, password) {
    return this.executeTrade(
      userId,
      {
        inputMint: config.tokens.USDC,
        outputMint: noOutcomeMint,
        amount: usdcAmount * 1e6,
        slippageBps
      },
      password
    );
  }
  /**
   * Sell outcome tokens back to USDC
   */
  async sellOutcome(userId, outcomeMint, tokenAmount, slippageBps, password) {
    return this.executeTrade(
      userId,
      {
        inputMint: outcomeMint,
        outputMint: config.tokens.USDC,
        amount: tokenAmount,
        slippageBps
      },
      password
    );
  }
  /**
   * Redeem winning outcome tokens after market settlement
   */
  async redeemWinnings(userId, outcomeMint, tokenAmount, password) {
    return this.executeTrade(
      userId,
      {
        inputMint: outcomeMint,
        outputMint: config.tokens.USDC,
        amount: tokenAmount,
        slippageBps: 10
        // Very low slippage for redemptions
      },
      password
    );
  }
  /**
   * Get swap instructions without executing (for custom signing flows)
   */
  async getSwapInstructions(params) {
    const response = await this.client.post("/swap/instructions", params);
    return response.data;
  }
  /**
   * Initialize a prediction market (tokenization)
   * This creates the YES/NO outcome tokens on-chain for an uninitialized market.
   * Requires a small SOL fee paid by the payer.
   * 
   * Reference: https://pond.dflow.net/swap-api-reference/prediction-market/prediction-market-init
   */
  async initializeMarket(userId, outcomeMint, password) {
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(userId);
    if (!walletInfo) {
      throw new Error("No wallet found for user");
    }
    if (walletInfo.type === "imported" && !password) {
      throw new Error("Password required for imported wallet transactions");
    }
    await this.throttle();
    const response = await this.client.get("/prediction-market-init", {
      params: {
        payer: walletInfo.publicKey,
        outcomeMint
      }
    });
    const { transaction, computeUnitLimit, lastValidBlockHeight } = response.data;
    if (!transaction) {
      throw new Error("No transaction returned - market may already be initialized");
    }
    const txBuffer = Buffer.from(transaction, "base64");
    let tx;
    try {
      tx = VersionedTransaction2.deserialize(txBuffer);
    } catch {
      tx = Transaction2.from(txBuffer);
    }
    const txSignature = await walletService.signAndSendTransaction(
      userId,
      tx,
      password
    );
    return {
      txSignature,
      success: true
    };
  }
  /**
   * Submit an intent-based swap (declarative mode)
   */
  async submitIntentSwap(params) {
    const response = await this.client.post("/intent/swap", params);
    return response.data;
  }
};
var tradeServiceInstance = null;
function getDFlowTradeService() {
  if (!tradeServiceInstance) {
    tradeServiceInstance = new DFlowTradeService();
  }
  return tradeServiceInstance;
}

// src/services/jupiter-swap.service.ts
import axios3 from "axios";
import {
  Connection as Connection3,
  VersionedTransaction as VersionedTransaction3
} from "@solana/web3.js";
var JupiterSwapService = class {
  client;
  connection;
  constructor() {
    this.client = axios3.create({
      baseURL: "https://quote-api.jup.ag/v6",
      timeout: 3e4,
      headers: {
        "Content-Type": "application/json"
      }
    });
    this.connection = new Connection3(config.solana.rpcUrl, config.solana.commitment);
  }
  /**
   * Get a swap quote from Jupiter
   */
  async getQuote(params) {
    const queryParams = {
      inputMint: params.inputMint,
      outputMint: params.outputMint,
      amount: params.amount,
      slippageBps: params.slippageBps || 50
      // Default 0.5%
    };
    if (params.swapMode) {
      queryParams.swapMode = params.swapMode;
    }
    if (params.onlyDirectRoutes !== void 0) {
      queryParams.onlyDirectRoutes = params.onlyDirectRoutes;
    }
    if (params.restrictIntermediateTokens !== void 0) {
      queryParams.restrictIntermediateTokens = params.restrictIntermediateTokens;
    }
    if (params.maxAccounts !== void 0) {
      queryParams.maxAccounts = params.maxAccounts;
    }
    const response = await this.client.get("/quote", { params: queryParams });
    return response.data;
  }
  /**
   * Get swap transaction from Jupiter
   */
  async getSwapTransaction(quote, userPublicKey, options) {
    const response = await this.client.post("/swap", {
      quoteResponse: quote,
      userPublicKey,
      wrapAndUnwrapSol: options?.wrapAndUnwrapSol ?? true,
      dynamicComputeUnitLimit: options?.dynamicComputeUnitLimit ?? true,
      prioritizationFeeLamports: options?.prioritizationFeeLamports ?? "auto"
    });
    return response.data;
  }
  /**
   * Execute a swap for a user
   */
  async executeSwap(userId, params, password) {
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(userId);
    if (!walletInfo) {
      throw new Error("No wallet found for user");
    }
    if (walletInfo.type === "imported" && !password) {
      throw new Error("Password required for imported wallet transactions");
    }
    const quote = await this.getQuote(params);
    const swapResponse = await this.getSwapTransaction(quote, walletInfo.publicKey);
    if (!swapResponse.swapTransaction) {
      throw new Error("No swap transaction returned - insufficient liquidity or invalid parameters");
    }
    const txBuffer = Buffer.from(swapResponse.swapTransaction, "base64");
    const transaction = VersionedTransaction3.deserialize(txBuffer);
    const txSignature = await walletService.signAndSendTransaction(
      userId,
      transaction,
      password
    );
    return {
      quote,
      txSignature,
      inputAmount: quote.inAmount,
      outputAmount: quote.outAmount,
      priceImpact: quote.priceImpactPct,
      explorerUrl: `https://solscan.io/tx/${txSignature}`
    };
  }
  /**
   * Helper: Swap SOL to USDC
   */
  async swapSolToUsdc(userId, solAmount, slippageBps, password) {
    const lamports = Math.floor(solAmount * 1e9);
    return this.executeSwap(
      userId,
      {
        inputMint: config.tokens.SOL,
        outputMint: config.tokens.USDC,
        amount: lamports,
        slippageBps: slippageBps || 50
      },
      password
    );
  }
  /**
   * Helper: Swap USDC to SOL
   */
  async swapUsdcToSol(userId, usdcAmount, slippageBps, password) {
    const microUsdc = Math.floor(usdcAmount * 1e6);
    return this.executeSwap(
      userId,
      {
        inputMint: config.tokens.USDC,
        outputMint: config.tokens.SOL,
        amount: microUsdc,
        slippageBps: slippageBps || 50
      },
      password
    );
  }
  /**
   * Get supported tokens list (subset of popular tokens)
   */
  getSupportedTokens() {
    return {
      SOL: {
        symbol: "SOL",
        mint: config.tokens.SOL,
        decimals: 9,
        name: "Wrapped SOL"
      },
      USDC: {
        symbol: "USDC",
        mint: config.tokens.USDC,
        decimals: 6,
        name: "USD Coin"
      },
      USDT: {
        symbol: "USDT",
        mint: config.tokens.USDT,
        decimals: 6,
        name: "Tether USD"
      }
    };
  }
};
var jupiterServiceInstance = null;
function getJupiterSwapService() {
  if (!jupiterServiceInstance) {
    jupiterServiceInstance = new JupiterSwapService();
  }
  return jupiterServiceInstance;
}

// src/mcp/tools.ts
var kalshiTools = [
  // ============================================
  // AUTHENTICATION & SETUP
  // ============================================
  {
    name: "kalshi_signup",
    description: "Create a new Kalshi account with a fresh Solana wallet. No access code required. Returns API credentials for all future requests.",
    inputSchema: {
      type: "object",
      properties: {
        externalId: {
          type: "string",
          description: "Your unique identifier (e.g., email, telegram:123456, or any unique string)"
        },
        keyName: {
          type: "string",
          description: "Optional friendly name for your API key"
        }
      },
      required: ["externalId"]
    }
  },
  {
    name: "kalshi_request_api_key",
    description: "Request a new API key for Kalshi trading. Requires an access code from Quantish. Returns credentials for MCP authentication.",
    inputSchema: {
      type: "object",
      properties: {
        accessCode: {
          type: "string",
          description: "Access code from Quantish (format: KALSHI-XXXX-XXXX-XXXX)"
        },
        externalId: {
          type: "string",
          description: "Your unique user identifier"
        },
        keyName: {
          type: "string",
          description: "Optional friendly name for this API key"
        }
      },
      required: ["accessCode", "externalId"]
    }
  },
  {
    name: "kalshi_setup_wallet",
    description: "Generate a new Solana wallet for Kalshi trading. The private key is encrypted and stored securely.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  {
    name: "kalshi_import_wallet",
    description: "Import an existing Solana wallet (e.g., from Phantom) for Kalshi trading. Send an encrypted bundle - we never see your raw private key.",
    inputSchema: {
      type: "object",
      properties: {
        encryptedKey: { type: "string", description: "Encrypted private key blob" },
        salt: { type: "string", description: "Salt used for encryption" },
        iv: { type: "string", description: "IV used for encryption" },
        publicKey: { type: "string", description: "Your Solana public key for verification" },
        version: { type: "string", description: "Bundle version (default: 1.0)" }
      },
      required: ["encryptedKey", "salt", "iv", "publicKey"]
    }
  },
  {
    name: "kalshi_get_wallet_import_instructions",
    description: "Get instructions for how to securely export your Solana wallet from Phantom/Solflare for import into Kalshi.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  // ============================================
  // WALLET INFO
  // ============================================
  {
    name: "kalshi_get_wallet_info",
    description: "Get your Kalshi wallet information including Solana public key and type (generated or imported).",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  {
    name: "kalshi_get_balances",
    description: "Get your SOL and USDC balances for Kalshi trading.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  {
    name: "kalshi_get_token_holdings",
    description: "Get all SPL token holdings including Kalshi prediction market positions.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  // ============================================
  // MARKET DISCOVERY
  // ============================================
  {
    name: "kalshi_search_markets",
    description: "Search for Kalshi prediction markets by keyword via DFlow. Returns markets with pagination support. Default limit is 10 results.",
    inputSchema: {
      type: "object",
      properties: {
        query: { type: "string", description: 'Search query (e.g., "bitcoin", "election", "weather")' },
        limit: { type: "number", description: "Max results per page (default 10, max 50)" },
        offset: { type: "number", description: "Number of results to skip for pagination (default 0)" },
        marketStatus: {
          type: "string",
          enum: ["active", "inactive", "finalized", "all"],
          description: "Filter by market status: active (tradeable), inactive (not yet open), finalized (settled), or all (default)"
        }
      },
      required: ["query"]
    }
  },
  {
    name: "kalshi_get_market",
    description: "Get details for a specific Kalshi market including prices and outcome mints.",
    inputSchema: {
      type: "object",
      properties: {
        ticker: { type: "string", description: "Kalshi market ticker (e.g., KXSB-26-BUF)" }
      },
      required: ["ticker"]
    }
  },
  {
    name: "kalshi_get_events",
    description: "Get list of Kalshi events with optional filters and pagination. Defaults to active (tradeable) markets only.",
    inputSchema: {
      type: "object",
      properties: {
        marketStatus: {
          type: "string",
          enum: ["active", "inactive", "finalized", "all"],
          description: "Filter by market status: active (tradeable, default), inactive (not yet open), finalized (settled), or all"
        },
        category: { type: "string", description: "Filter by category" },
        limit: { type: "number", description: "Max results per page (default 10, max 50)" },
        cursor: { type: "string", description: "Pagination cursor from previous response" }
      },
      required: []
    }
  },
  {
    name: "kalshi_get_live_data",
    description: "Get live pricing data for a Kalshi market.",
    inputSchema: {
      type: "object",
      properties: {
        marketTicker: { type: "string", description: "Kalshi market ticker" }
      },
      required: ["marketTicker"]
    }
  },
  // ============================================
  // MARKET INITIALIZATION & STATUS
  // ============================================
  {
    name: "kalshi_check_market_initialization",
    description: "Check if a Kalshi market is initialized (tokenized) on-chain. Uninitialized markets require a small SOL fee to be paid on first trade.",
    inputSchema: {
      type: "object",
      properties: {
        ticker: { type: "string", description: "Kalshi market ticker" },
        settlementMint: {
          type: "string",
          description: "Settlement token mint (default: USDC)",
          default: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"
        }
      },
      required: ["ticker"]
    }
  },
  {
    name: "kalshi_initialize_market",
    description: "Initialize (tokenize) a Kalshi market on-chain. This creates the YES/NO outcome tokens. Requires SOL for the transaction fee. Only needed if market is not yet initialized.",
    inputSchema: {
      type: "object",
      properties: {
        outcomeMint: {
          type: "string",
          description: "Either YES or NO outcome mint address for the market (use kalshi_check_market_initialization to get this)"
        },
        password: { type: "string", description: "Password (required for imported wallets only)" }
      },
      required: ["outcomeMint"]
    }
  },
  {
    name: "kalshi_check_redemption_status",
    description: "Check if a Kalshi market is settled and positions can be redeemed for stablecoins. Returns winning/losing side and redemption availability.",
    inputSchema: {
      type: "object",
      properties: {
        ticker: { type: "string", description: "Kalshi market ticker" },
        settlementMint: {
          type: "string",
          description: "Settlement token mint (default: USDC)",
          default: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"
        }
      },
      required: ["ticker"]
    }
  },
  {
    name: "kalshi_get_market_by_mint",
    description: "Look up a Kalshi market by outcome token mint address. Useful when you have a token in your wallet and want to find the associated market.",
    inputSchema: {
      type: "object",
      properties: {
        mintAddress: { type: "string", description: "Outcome token mint address (Solana SPL)" }
      },
      required: ["mintAddress"]
    }
  },
  // ============================================
  // TRADING
  // ============================================
  {
    name: "kalshi_get_quote",
    description: "Get a quote for buying/selling Kalshi prediction market tokens via DFlow.",
    inputSchema: {
      type: "object",
      properties: {
        inputMint: { type: "string", description: "Token to spend (USDC mint or outcome token)" },
        outputMint: { type: "string", description: "Token to receive (outcome token or USDC)" },
        amount: { type: "number", description: "Amount in smallest units (e.g., 1000000 for 1 USDC)" },
        slippageBps: { type: "number", description: "Slippage tolerance in basis points (default 100 = 1%)" }
      },
      required: ["inputMint", "outputMint", "amount"]
    }
  },
  {
    name: "kalshi_buy_yes",
    description: "Buy YES outcome tokens for a Kalshi market.",
    inputSchema: {
      type: "object",
      properties: {
        marketTicker: { type: "string", description: "Kalshi market ticker" },
        yesOutcomeMint: { type: "string", description: "YES outcome token mint address (Solana SPL)" },
        usdcAmount: { type: "number", description: "Amount of USDC to spend (e.g., 10 for $10)" },
        slippageBps: { type: "number", description: "Slippage tolerance (default 100 = 1%)" },
        password: { type: "string", description: "Password (required for imported wallets only)" }
      },
      required: ["marketTicker", "yesOutcomeMint", "usdcAmount"]
    }
  },
  {
    name: "kalshi_buy_no",
    description: "Buy NO outcome tokens for a Kalshi market.",
    inputSchema: {
      type: "object",
      properties: {
        marketTicker: { type: "string", description: "Kalshi market ticker" },
        noOutcomeMint: { type: "string", description: "NO outcome token mint address (Solana SPL)" },
        usdcAmount: { type: "number", description: "Amount of USDC to spend" },
        slippageBps: { type: "number", description: "Slippage tolerance (default 100 = 1%)" },
        password: { type: "string", description: "Password (required for imported wallets only)" }
      },
      required: ["marketTicker", "noOutcomeMint", "usdcAmount"]
    }
  },
  {
    name: "kalshi_sell_position",
    description: "Sell Kalshi outcome tokens back to USDC.",
    inputSchema: {
      type: "object",
      properties: {
        outcomeMint: { type: "string", description: "Outcome token mint to sell" },
        tokenAmount: { type: "number", description: "Number of tokens to sell" },
        slippageBps: { type: "number", description: "Slippage tolerance (default 100 = 1%)" },
        password: { type: "string", description: "Password (required for imported wallets only)" }
      },
      required: ["outcomeMint", "tokenAmount"]
    }
  },
  {
    name: "kalshi_redeem_winnings",
    description: "Redeem winning Kalshi outcome tokens after market settlement.",
    inputSchema: {
      type: "object",
      properties: {
        outcomeMint: { type: "string", description: "Winning outcome token mint" },
        tokenAmount: { type: "number", description: "Number of tokens to redeem" },
        password: { type: "string", description: "Password (required for imported wallets only)" }
      },
      required: ["outcomeMint", "tokenAmount"]
    }
  },
  // ============================================
  // POSITIONS & HISTORY
  // ============================================
  {
    name: "kalshi_get_positions",
    description: "Get your current Kalshi prediction market positions.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  {
    name: "kalshi_get_orders",
    description: "Get your Kalshi order history.",
    inputSchema: {
      type: "object",
      properties: {
        status: { type: "string", enum: ["PENDING", "SUBMITTED", "FILLED", "CANCELLED", "FAILED"] },
        limit: { type: "number", description: "Max results" }
      },
      required: []
    }
  },
  // ============================================
  // API KEY MANAGEMENT
  // ============================================
  {
    name: "kalshi_list_api_keys",
    description: "List all your Kalshi API keys (without exposing the actual keys).",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  {
    name: "kalshi_create_additional_api_key",
    description: "Create an additional API key for your Kalshi account.",
    inputSchema: {
      type: "object",
      properties: {
        name: { type: "string", description: "Friendly name for this key" }
      },
      required: []
    }
  },
  {
    name: "kalshi_revoke_api_key",
    description: "Revoke one of your Kalshi API keys.",
    inputSchema: {
      type: "object",
      properties: {
        keyId: { type: "string", description: "ID of the key to revoke" }
      },
      required: ["keyId"]
    }
  },
  // ============================================
  // WALLET STATUS & DIAGNOSTICS
  // ============================================
  {
    name: "kalshi_get_wallet_status",
    description: "Get comprehensive wallet status including balances, token holdings, and connection health.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  {
    name: "kalshi_get_deposit_address",
    description: "Get your Solana wallet address for depositing SOL and USDC.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  {
    name: "kalshi_export_private_key",
    description: "Export your Solana wallet private key. Returns the raw private key for backup. WARNING: Anyone with this key controls your wallet.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  {
    name: "kalshi_import_private_key",
    description: "Import an existing Solana private key directly (simpler than encrypted import). Creates a new wallet linked to your account.",
    inputSchema: {
      type: "object",
      properties: {
        externalId: {
          type: "string",
          description: "Your unique identifier (e.g., telegram:123456789)"
        },
        privateKey: {
          type: "string",
          description: "Base58-encoded Solana private key"
        },
        keyName: {
          type: "string",
          description: "Optional friendly name for the API key"
        }
      },
      required: ["externalId", "privateKey"]
    }
  },
  // ============================================
  // TOKEN SWAPS (via Jupiter)
  // ============================================
  {
    name: "kalshi_get_swap_quote",
    description: "Get a quote for swapping tokens via Jupiter Aggregator. Use this for SOL \u2194 USDC swaps and other token swaps (not prediction market trading).",
    inputSchema: {
      type: "object",
      properties: {
        inputMint: {
          type: "string",
          description: 'Input token mint address. Use "SOL" for native SOL, "USDC" for USDC.'
        },
        outputMint: {
          type: "string",
          description: 'Output token mint address. Use "SOL" for native SOL, "USDC" for USDC.'
        },
        amount: {
          type: "number",
          description: "Amount to swap in human-readable units (e.g., 1.5 for 1.5 SOL or 100 for $100 USDC)"
        },
        slippageBps: {
          type: "number",
          description: "Slippage tolerance in basis points (default 50 = 0.5%)"
        }
      },
      required: ["inputMint", "outputMint", "amount"]
    }
  },
  {
    name: "kalshi_execute_swap",
    description: "Execute a token swap via Jupiter Aggregator. Use this to swap SOL to USDC, USDC to SOL, or other token pairs.",
    inputSchema: {
      type: "object",
      properties: {
        inputMint: {
          type: "string",
          description: 'Input token: "SOL", "USDC", or mint address'
        },
        outputMint: {
          type: "string",
          description: 'Output token: "SOL", "USDC", or mint address'
        },
        amount: {
          type: "number",
          description: "Amount to swap in human-readable units"
        },
        slippageBps: {
          type: "number",
          description: "Slippage tolerance in basis points (default 50 = 0.5%)"
        },
        password: {
          type: "string",
          description: "Password (required for imported wallets only)"
        }
      },
      required: ["inputMint", "outputMint", "amount"]
    }
  },
  {
    name: "kalshi_swap_sol_to_usdc",
    description: "Swap SOL to USDC. Convenience method for the most common swap.",
    inputSchema: {
      type: "object",
      properties: {
        solAmount: {
          type: "number",
          description: "Amount of SOL to swap (e.g., 0.5 for 0.5 SOL)"
        },
        slippageBps: {
          type: "number",
          description: "Slippage tolerance in basis points (default 50 = 0.5%)"
        },
        password: {
          type: "string",
          description: "Password (required for imported wallets only)"
        }
      },
      required: ["solAmount"]
    }
  },
  {
    name: "kalshi_swap_usdc_to_sol",
    description: "Swap USDC to SOL. Convenience method for getting SOL for transaction fees.",
    inputSchema: {
      type: "object",
      properties: {
        usdcAmount: {
          type: "number",
          description: "Amount of USDC to swap (e.g., 10 for $10)"
        },
        slippageBps: {
          type: "number",
          description: "Slippage tolerance in basis points (default 50 = 0.5%)"
        },
        password: {
          type: "string",
          description: "Password (required for imported wallets only)"
        }
      },
      required: ["usdcAmount"]
    }
  },
  // ============================================
  // SEND TOKENS (Withdrawals)
  // ============================================
  {
    name: "kalshi_send_sol",
    description: "Send SOL to another Solana wallet. Use this to withdraw SOL to an external wallet.",
    inputSchema: {
      type: "object",
      properties: {
        toAddress: {
          type: "string",
          description: "Destination Solana wallet address (base58)"
        },
        amount: {
          type: "number",
          description: "Amount of SOL to send (e.g., 0.1 for 0.1 SOL)"
        },
        password: {
          type: "string",
          description: "Password (required for imported wallets only)"
        }
      },
      required: ["toAddress", "amount"]
    }
  },
  {
    name: "kalshi_send_usdc",
    description: "Send USDC to another Solana wallet. Use this to withdraw USDC to an external wallet.",
    inputSchema: {
      type: "object",
      properties: {
        toAddress: {
          type: "string",
          description: "Destination Solana wallet address (base58)"
        },
        amount: {
          type: "number",
          description: "Amount of USDC to send (e.g., 50 for $50)"
        },
        password: {
          type: "string",
          description: "Password (required for imported wallets only)"
        }
      },
      required: ["toAddress", "amount"]
    }
  },
  {
    name: "kalshi_send_token",
    description: "Send any SPL token to another Solana wallet. Use for tokens other than SOL/USDC.",
    inputSchema: {
      type: "object",
      properties: {
        toAddress: {
          type: "string",
          description: "Destination Solana wallet address (base58)"
        },
        mintAddress: {
          type: "string",
          description: "SPL token mint address"
        },
        amount: {
          type: "number",
          description: "Amount to send in token units"
        },
        decimals: {
          type: "number",
          description: "Token decimals (e.g., 6 for USDC, 9 for SOL)"
        },
        password: {
          type: "string",
          description: "Password (required for imported wallets only)"
        }
      },
      required: ["toAddress", "mintAddress", "amount", "decimals"]
    }
  },
  // ============================================
  // REDEMPTION (Additional helpers)
  // ============================================
  {
    name: "kalshi_get_redeemable_positions",
    description: "Get all your positions that can be redeemed (from settled/finalized markets). Returns winning positions ready to claim.",
    inputSchema: {
      type: "object",
      properties: {},
      required: []
    }
  },
  {
    name: "kalshi_redeem_all_positions",
    description: "Redeem all winning positions at once. Swaps all redeemable outcome tokens back to USDC.",
    inputSchema: {
      type: "object",
      properties: {
        password: {
          type: "string",
          description: "Password (required for imported wallets only)"
        }
      },
      required: []
    }
  }
];
async function executeTool(name, args, context) {
  const prisma2 = getPrismaClient();
  const { userId } = context;
  if (name === "kalshi_signup") {
    const { externalId, keyName } = args;
    if (!externalId) {
      throw new Error("externalId is required");
    }
    let existingUser = await prisma2.user.findUnique({
      where: { externalId }
    });
    if (existingUser) {
      throw new Error("User with this externalId already exists. Use a different externalId or use kalshi_request_api_key with an access code.");
    }
    const newUser = await prisma2.user.create({
      data: {
        externalId
      }
    });
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.generateWallet(newUser.id);
    const publicKey = walletInfo.publicKey;
    const apiKeyService = getApiKeyService();
    const result = await apiKeyService.createApiKey(newUser.id, keyName || "Default");
    await prisma2.activityLog.create({
      data: {
        userId: newUser.id,
        action: "SIGNUP",
        resource: "user",
        details: {
          timestamp: (/* @__PURE__ */ new Date()).toISOString(),
          method: "self-service"
        }
      }
    });
    return {
      success: true,
      message: "Account created successfully! Save your API key - it cannot be recovered.",
      apiKey: result.apiKey,
      apiSecret: result.apiSecret,
      walletAddress: publicKey,
      nextSteps: [
        "Save your API key securely - it will not be shown again",
        `Fund your wallet with SOL and USDC: ${publicKey}`,
        "Use your API key in the x-api-key header for all future requests"
      ]
    };
  }
  if (name === "kalshi_request_api_key") {
    const { accessCode, externalId, keyName } = args;
    const accessCodeService = getAccessCodeService();
    const validation = await accessCodeService.validateAndUseAccessCode(accessCode, externalId);
    if (!validation.isValid) {
      throw new Error(validation.message);
    }
    let user2 = await prisma2.user.findUnique({
      where: { externalId }
    });
    if (!user2) {
      user2 = await prisma2.user.create({
        data: { externalId }
      });
    }
    const apiKeyService = getApiKeyService();
    const result = await apiKeyService.createApiKey(user2.id, keyName);
    return {
      message: "API key created successfully",
      apiKey: result.apiKey,
      apiSecret: result.apiSecret,
      keyPrefix: result.keyPrefix,
      userId: user2.id,
      nextSteps: {
        cursorMcp: {
          description: "Add this to your Cursor MCP config (~/.cursor/mcp.json):",
          config: {
            quantish_kalshi: {
              url: "https://kalshi-mcp-server-production.up.railway.app/mcp",
              headers: {
                "x-api-key": result.apiKey
              }
            }
          }
        }
      }
    };
  }
  if (name === "kalshi_get_wallet_import_instructions") {
    return {
      instructions: getWalletExportInstructions(),
      encryptionExample: `
// Run this in Node.js on YOUR machine (not on our servers)
const crypto = require('crypto');

function encryptWallet(privateKeyBase58, password) {
  const salt = crypto.randomBytes(32);
  const derivedKey = crypto.pbkdf2Sync(password, salt, 100000, 32, 'sha512');
  const iv = crypto.randomBytes(16);
  const cipher = crypto.createCipheriv('aes-256-gcm', derivedKey, iv);
  
  let encrypted = cipher.update(privateKeyBase58, 'utf8', 'hex');
  encrypted += cipher.final('hex');
  const authTag = cipher.getAuthTag().toString('hex');
  
  return {
    encryptedKey: encrypted + ':' + authTag,
    salt: salt.toString('hex'),
    iv: iv.toString('hex'),
    publicKey: 'YOUR_PUBLIC_KEY_HERE',
    version: '1.0'
  };
}
`
    };
  }
  if (name === "kalshi_search_markets") {
    const marketService2 = getDFlowMarketService();
    const limit = Math.min(args.limit || 10, 50);
    const offset = args.offset || 0;
    const allEvents = await marketService2.searchEvents(
      args.query,
      offset + limit + 1,
      args.marketStatus || "all"
    );
    const paginatedEvents = allEvents.slice(offset, offset + limit);
    const hasMore = allEvents.length > offset + limit;
    return {
      events: paginatedEvents,
      pagination: {
        offset,
        limit,
        returned: paginatedEvents.length,
        hasMore,
        nextOffset: hasMore ? offset + limit : null
      },
      note: 'Markets have statuses: active (tradeable), inactive (not yet open), finalized (settled). Only "active" markets can be traded.'
    };
  }
  if (name === "kalshi_get_market") {
    const marketService2 = getDFlowMarketService();
    const market = await marketService2.getMarket(args.ticker);
    if (!market) {
      return {
        error: `Market not found: ${args.ticker}`,
        suggestion: "Use kalshi_search_markets or kalshi_get_events to find valid market tickers"
      };
    }
    return { market };
  }
  if (name === "kalshi_get_events") {
    const marketService2 = getDFlowMarketService();
    const limit = Math.min(args.limit || 10, 50);
    const result = await marketService2.getEvents({
      marketStatus: args.marketStatus || "active",
      // Default to active markets only
      category: args.category,
      limit,
      cursor: args.cursor,
      withNestedMarkets: true
    });
    return {
      events: result.events,
      pagination: {
        limit,
        returned: result.events.length,
        cursor: result.cursor || null,
        hasMore: !!result.cursor
      },
      note: "Markets have statuses: active (tradeable), inactive (not yet open), finalized (settled). Use cursor for pagination."
    };
  }
  if (name === "kalshi_get_live_data") {
    const marketService2 = getDFlowMarketService();
    const liveData = await marketService2.getLiveData(args.marketTicker);
    if (!liveData) {
      return {
        error: `Market not found: ${args.marketTicker}`,
        suggestion: "Use kalshi_search_markets or kalshi_get_events to find valid market tickers"
      };
    }
    return { liveData };
  }
  if (name === "kalshi_check_market_initialization") {
    const marketService2 = getDFlowMarketService();
    const result = await marketService2.checkMarketInitialization(
      args.ticker,
      args.settlementMint
    );
    return result;
  }
  if (name === "kalshi_check_redemption_status") {
    const marketService2 = getDFlowMarketService();
    const result = await marketService2.checkRedemptionStatus(args.ticker);
    return result;
  }
  if (name === "kalshi_get_market_by_mint") {
    const marketService2 = getDFlowMarketService();
    const market = await marketService2.getMarketByMint(args.mintAddress);
    if (!market) {
      return {
        error: `No market found for mint: ${args.mintAddress}`
      };
    }
    return { market };
  }
  if (!userId) {
    throw new Error("Authentication required. Please provide a valid API key.");
  }
  const user = await prisma2.user.findUnique({
    where: { id: userId }
  });
  if (!user) {
    throw new Error("User not found");
  }
  if (name === "kalshi_setup_wallet") {
    const walletService = getSolanaWalletService();
    const existingWallet = await walletService.getWalletInfo(userId);
    if (existingWallet) {
      return {
        message: "Wallet already exists",
        wallet: existingWallet
      };
    }
    const wallet = await walletService.generateWallet(userId);
    return {
      message: "Solana wallet generated successfully",
      wallet,
      note: "Your private key is encrypted and stored securely. Fund this address with SOL and USDC to start trading."
    };
  }
  if (name === "kalshi_import_wallet") {
    const bundle = {
      encryptedKey: args.encryptedKey,
      salt: args.salt,
      iv: args.iv,
      publicKey: args.publicKey,
      version: args.version || "1.0"
    };
    if (!verifyWalletImportBundle(bundle)) {
      throw new Error("Invalid wallet import bundle format");
    }
    const walletService = getSolanaWalletService();
    const wallet = await walletService.importWallet(userId, bundle);
    return {
      message: "Wallet imported successfully",
      wallet,
      note: "You will need to provide your password each time you make a transaction."
    };
  }
  if (name === "kalshi_get_wallet_info") {
    const walletService = getSolanaWalletService();
    const wallet = await walletService.getWalletInfo(userId);
    if (!wallet) {
      return { message: "No wallet found. Use kalshi_setup_wallet or kalshi_import_wallet first." };
    }
    return { wallet };
  }
  if (name === "kalshi_get_balances") {
    const walletService = getSolanaWalletService();
    const wallet = await walletService.getWalletInfo(userId);
    if (!wallet) {
      throw new Error("No wallet found");
    }
    const balances = await walletService.getBalances(wallet.publicKey);
    return {
      publicKey: wallet.publicKey,
      balances
    };
  }
  if (name === "kalshi_get_token_holdings") {
    const walletService = getSolanaWalletService();
    const wallet = await walletService.getWalletInfo(userId);
    if (!wallet) {
      throw new Error("No wallet found");
    }
    const tokens = await walletService.getAllTokenAccounts(wallet.publicKey);
    return {
      publicKey: wallet.publicKey,
      tokens
    };
  }
  if (name === "kalshi_initialize_market") {
    const tradeService = getDFlowTradeService();
    try {
      const result = await tradeService.initializeMarket(
        userId,
        args.outcomeMint,
        args.password
      );
      return {
        success: true,
        txSignature: result.txSignature,
        message: "Market initialized successfully. The YES/NO outcome tokens are now created on-chain.",
        explorerUrl: `https://solscan.io/tx/${result.txSignature}`
      };
    } catch (error) {
      if (error.message?.includes("already be initialized")) {
        return {
          success: true,
          message: "Market is already initialized. No action needed."
        };
      }
      throw error;
    }
  }
  if (name === "kalshi_check_redemption_status") {
    const marketService2 = getDFlowMarketService();
    const result = await marketService2.checkRedemptionStatus(
      args.ticker,
      args.settlementMint
    );
    return result;
  }
  if (name === "kalshi_get_market_by_mint") {
    const marketService2 = getDFlowMarketService();
    const market = await marketService2.getMarketByMint(args.mintAddress);
    if (!market) {
      return {
        error: "Market not found for this mint address",
        suggestion: "This mint may not be a Kalshi outcome token."
      };
    }
    return { market };
  }
  if (name === "kalshi_get_quote") {
    const walletService = getSolanaWalletService();
    const wallet = await walletService.getWalletInfo(userId);
    if (!wallet) {
      throw new Error("No wallet found");
    }
    const tradeService = getDFlowTradeService();
    const quote = await tradeService.getQuote({
      userPublicKey: wallet.publicKey,
      inputMint: args.inputMint,
      outputMint: args.outputMint,
      amount: args.amount,
      slippageBps: args.slippageBps
    });
    return { quote };
  }
  if (name === "kalshi_buy_yes") {
    const tradeService = getDFlowTradeService();
    const result = await tradeService.buyYes(
      userId,
      args.marketTicker,
      args.yesOutcomeMint,
      args.usdcAmount,
      args.slippageBps,
      args.password
    );
    await prisma2.order.create({
      data: {
        userId,
        eventTicker: args.marketTicker,
        marketTicker: args.marketTicker,
        outcomeMint: args.yesOutcomeMint,
        side: "BUY",
        amount: args.usdcAmount,
        txSignature: result.txSignature,
        status: result.status === "success" ? "FILLED" : "SUBMITTED"
      }
    });
    return {
      message: "Buy YES order executed",
      ...result
    };
  }
  if (name === "kalshi_buy_no") {
    const tradeService = getDFlowTradeService();
    const result = await tradeService.buyNo(
      userId,
      args.marketTicker,
      args.noOutcomeMint,
      args.usdcAmount,
      args.slippageBps,
      args.password
    );
    await prisma2.order.create({
      data: {
        userId,
        eventTicker: args.marketTicker,
        marketTicker: args.marketTicker,
        outcomeMint: args.noOutcomeMint,
        side: "BUY",
        amount: args.usdcAmount,
        txSignature: result.txSignature,
        status: result.status === "success" ? "FILLED" : "SUBMITTED"
      }
    });
    return {
      message: "Buy NO order executed",
      ...result
    };
  }
  if (name === "kalshi_sell_position") {
    const tradeService = getDFlowTradeService();
    const result = await tradeService.sellOutcome(
      userId,
      args.outcomeMint,
      args.tokenAmount,
      args.slippageBps,
      args.password
    );
    await prisma2.order.create({
      data: {
        userId,
        eventTicker: "UNKNOWN",
        marketTicker: "UNKNOWN",
        outcomeMint: args.outcomeMint,
        side: "SELL",
        amount: args.tokenAmount,
        txSignature: result.txSignature,
        status: result.status === "success" ? "FILLED" : "SUBMITTED"
      }
    });
    return {
      message: "Sell order executed",
      ...result
    };
  }
  if (name === "kalshi_redeem_winnings") {
    const tradeService = getDFlowTradeService();
    const result = await tradeService.redeemWinnings(
      userId,
      args.outcomeMint,
      args.tokenAmount,
      args.password
    );
    return {
      message: "Redemption executed",
      ...result
    };
  }
  if (name === "kalshi_get_positions") {
    const positions = await prisma2.position.findMany({
      where: { userId },
      orderBy: { updatedAt: "desc" }
    });
    return { positions };
  }
  if (name === "kalshi_get_orders") {
    const orders = await prisma2.order.findMany({
      where: {
        userId,
        ...args.status ? { status: args.status } : {}
      },
      orderBy: { createdAt: "desc" },
      take: args.limit || 50
    });
    return { orders };
  }
  if (name === "kalshi_list_api_keys") {
    const apiKeyService = getApiKeyService();
    const keys = await apiKeyService.listApiKeys(userId);
    return { keys };
  }
  if (name === "kalshi_create_additional_api_key") {
    const apiKeyService = getApiKeyService();
    const result = await apiKeyService.createApiKey(userId, args.name);
    return {
      message: "Additional API key created",
      apiKey: result.apiKey,
      apiSecret: result.apiSecret,
      keyPrefix: result.keyPrefix
    };
  }
  if (name === "kalshi_revoke_api_key") {
    const apiKeyService = getApiKeyService();
    const success = await apiKeyService.revokeApiKey(userId, args.keyId);
    return {
      success,
      message: success ? "API key revoked" : "API key not found"
    };
  }
  if (name === "kalshi_get_wallet_status") {
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(userId);
    if (!walletInfo) {
      return {
        hasWallet: false,
        message: "No wallet found. Use kalshi_setup_wallet or kalshi_import_wallet first."
      };
    }
    const balances = await walletService.getBalances(walletInfo.publicKey);
    const tokens = await walletService.getAllTokenAccounts(walletInfo.publicKey);
    const predictionMarketTokens = tokens.filter(
      (t) => t.mint !== "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v" && // Not USDC
      t.mint !== "So11111111111111111111111111111111111111112"
      // Not SOL
    );
    return {
      hasWallet: true,
      walletType: walletInfo.type,
      publicKey: walletInfo.publicKey,
      balances: {
        sol: balances.sol,
        usdc: balances.usdc
      },
      predictionMarketTokenCount: predictionMarketTokens.length,
      status: balances.sol > 1e-3 && balances.usdc > 0 ? "READY" : "NEEDS_FUNDING",
      message: balances.sol < 1e-3 ? "Low SOL balance - need SOL for transaction fees" : balances.usdc === 0 ? "No USDC balance - deposit USDC to start trading" : "Wallet ready for trading"
    };
  }
  if (name === "kalshi_get_deposit_address") {
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(userId);
    if (!walletInfo) {
      return {
        error: "No wallet found. Use kalshi_setup_wallet or kalshi_import_wallet first."
      };
    }
    return {
      publicKey: walletInfo.publicKey,
      walletType: walletInfo.type,
      network: "Solana Mainnet",
      acceptedTokens: [
        { name: "SOL", description: "For transaction fees", mint: "Native SOL" },
        { name: "USDC", description: "For trading", mint: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v" }
      ],
      instructions: `Send SOL and USDC to: ${walletInfo.publicKey}`
    };
  }
  if (name === "kalshi_export_private_key") {
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(userId);
    if (!walletInfo) {
      throw new Error("No wallet found");
    }
    if (walletInfo.type === "imported") {
      return {
        error: "Cannot export imported wallet",
        message: "Imported wallets are encrypted with your password. We never have access to your private key."
      };
    }
    const keypair = await walletService.getKeypairForSigning(userId);
    const privateKeyBase58 = __require("bs58").encode(keypair.secretKey);
    await prisma2.activityLog.create({
      data: {
        userId,
        action: "PRIVATE_KEY_EXPORT",
        resource: "wallet",
        details: {
          timestamp: (/* @__PURE__ */ new Date()).toISOString(),
          note: "Raw private key exported"
        }
      }
    });
    return {
      privateKey: privateKeyBase58,
      publicKey: walletInfo.publicKey,
      warning: "\u26A0\uFE0F SECURITY WARNING: Anyone with this private key controls your wallet. Store it securely and never share it."
    };
  }
  if (name === "kalshi_import_private_key") {
    const { externalId, privateKey, keyName } = args;
    if (!externalId) {
      throw new Error("externalId is required");
    }
    if (!privateKey) {
      throw new Error("privateKey is required");
    }
    let keypair;
    try {
      const bs583 = __require("bs58");
      const privateKeyBytes = bs583.decode(privateKey);
      const { Keypair: Keypair3 } = __require("@solana/web3.js");
      keypair = Keypair3.fromSecretKey(privateKeyBytes);
    } catch (e) {
      throw new Error("Invalid private key format. Must be Base58-encoded Solana private key.");
    }
    const publicKey = keypair.publicKey.toBase58();
    let existingUser = await prisma2.user.findUnique({
      where: { externalId }
    });
    if (existingUser) {
      throw new Error("User with this externalId already exists. Use a different externalId.");
    }
    const encryption = getEncryptionService();
    const encryptedPrivateKey = encryption.encrypt(privateKey);
    const newUser = await prisma2.user.create({
      data: {
        externalId,
        solanaPublicKey: publicKey,
        encryptedPrivateKey
      }
    });
    const apiKeyService = getApiKeyService();
    const keyResult = await apiKeyService.createApiKey(newUser.id, keyName || "Imported Wallet");
    await prisma2.activityLog.create({
      data: {
        userId: newUser.id,
        action: "PRIVATE_KEY_IMPORT",
        resource: "wallet",
        details: {
          timestamp: (/* @__PURE__ */ new Date()).toISOString(),
          keyName: keyName || "Imported Wallet"
        }
      }
    });
    return {
      success: true,
      apiKey: keyResult.apiKey,
      apiSecret: keyResult.apiSecret,
      publicKey,
      message: "Private key imported successfully. Save your API key - it cannot be recovered!"
    };
  }
  const resolveMint = (input) => {
    const upper = input.toUpperCase();
    if (upper === "SOL") return config.tokens.SOL;
    if (upper === "USDC") return config.tokens.USDC;
    if (upper === "USDT") return config.tokens.USDT;
    return input;
  };
  const getDecimals = (mint) => {
    if (mint === config.tokens.SOL) return 9;
    if (mint === config.tokens.USDC) return 6;
    if (mint === config.tokens.USDT) return 6;
    return 6;
  };
  if (name === "kalshi_get_swap_quote") {
    const jupiterService = getJupiterSwapService();
    const inputMint = resolveMint(args.inputMint);
    const outputMint = resolveMint(args.outputMint);
    const decimals = getDecimals(inputMint);
    const rawAmount = Math.floor(args.amount * Math.pow(10, decimals));
    const quote = await jupiterService.getQuote({
      inputMint,
      outputMint,
      amount: rawAmount,
      slippageBps: args.slippageBps
    });
    const outputDecimals = getDecimals(outputMint);
    return {
      quote: {
        inputMint: quote.inputMint,
        outputMint: quote.outputMint,
        inputAmount: args.amount,
        outputAmount: Number(quote.outAmount) / Math.pow(10, outputDecimals),
        priceImpact: quote.priceImpactPct,
        slippageBps: quote.slippageBps,
        routes: quote.routePlan?.map((r) => r.swapInfo.label).join(" \u2192 ")
      },
      note: "Use kalshi_execute_swap to execute this swap."
    };
  }
  if (name === "kalshi_execute_swap") {
    const jupiterService = getJupiterSwapService();
    const inputMint = resolveMint(args.inputMint);
    const outputMint = resolveMint(args.outputMint);
    const decimals = getDecimals(inputMint);
    const rawAmount = Math.floor(args.amount * Math.pow(10, decimals));
    const result = await jupiterService.executeSwap(
      userId,
      {
        inputMint,
        outputMint,
        amount: rawAmount,
        slippageBps: args.slippageBps
      },
      args.password
    );
    const outputDecimals = getDecimals(outputMint);
    return {
      success: true,
      txSignature: result.txSignature,
      inputAmount: args.amount,
      outputAmount: Number(result.outputAmount) / Math.pow(10, outputDecimals),
      priceImpact: result.priceImpact,
      explorerUrl: result.explorerUrl
    };
  }
  if (name === "kalshi_swap_sol_to_usdc") {
    const jupiterService = getJupiterSwapService();
    const result = await jupiterService.swapSolToUsdc(
      userId,
      args.solAmount,
      args.slippageBps,
      args.password
    );
    return {
      success: true,
      txSignature: result.txSignature,
      solSpent: args.solAmount,
      usdcReceived: Number(result.outputAmount) / 1e6,
      priceImpact: result.priceImpact,
      explorerUrl: result.explorerUrl
    };
  }
  if (name === "kalshi_swap_usdc_to_sol") {
    const jupiterService = getJupiterSwapService();
    const result = await jupiterService.swapUsdcToSol(
      userId,
      args.usdcAmount,
      args.slippageBps,
      args.password
    );
    return {
      success: true,
      txSignature: result.txSignature,
      usdcSpent: args.usdcAmount,
      solReceived: Number(result.outputAmount) / 1e9,
      priceImpact: result.priceImpact,
      explorerUrl: result.explorerUrl
    };
  }
  if (name === "kalshi_send_sol") {
    const walletService = getSolanaWalletService();
    const result = await walletService.sendSol(
      userId,
      args.toAddress,
      args.amount,
      args.password
    );
    return {
      success: true,
      ...result,
      message: `Successfully sent ${args.amount} SOL to ${args.toAddress}`
    };
  }
  if (name === "kalshi_send_usdc") {
    const walletService = getSolanaWalletService();
    const result = await walletService.sendUsdc(
      userId,
      args.toAddress,
      args.amount,
      args.password
    );
    return {
      success: true,
      ...result,
      message: `Successfully sent $${args.amount} USDC to ${args.toAddress}`
    };
  }
  if (name === "kalshi_send_token") {
    const walletService = getSolanaWalletService();
    const result = await walletService.sendToken(
      userId,
      args.toAddress,
      args.mintAddress,
      args.amount,
      args.decimals,
      args.password
    );
    return {
      success: true,
      ...result,
      message: `Successfully sent ${args.amount} tokens to ${args.toAddress}`
    };
  }
  if (name === "kalshi_get_redeemable_positions") {
    const walletService = getSolanaWalletService();
    const marketService2 = getDFlowMarketService();
    const walletInfo = await walletService.getWalletInfo(userId);
    if (!walletInfo) {
      return { positions: [], message: "No wallet found" };
    }
    const tokens = await walletService.getAllTokenAccounts(walletInfo.publicKey);
    const potentialPositions = tokens.filter(
      (t) => Number(t.balance) > 0 && t.mint !== config.tokens.USDC && t.mint !== config.tokens.SOL && t.mint !== config.tokens.USDT
    );
    const redeemablePositions = [];
    for (const position of potentialPositions) {
      try {
        const market = await marketService2.getMarketByMint(position.mint);
        if (market) {
          const redemptionStatus = await marketService2.checkRedemptionStatus(market.ticker);
          if (redemptionStatus.canRedeem) {
            redeemablePositions.push({
              mint: position.mint,
              balance: Number(position.balance) / Math.pow(10, position.decimals),
              market: market.ticker,
              marketTitle: market.title,
              outcome: redemptionStatus.winningOutcome,
              settlementValue: redemptionStatus.settlementValue
            });
          }
        }
      } catch {
      }
    }
    return {
      redeemablePositions,
      count: redeemablePositions.length,
      note: redeemablePositions.length > 0 ? "Use kalshi_redeem_winnings or kalshi_redeem_all_positions to claim these." : "No redeemable positions found."
    };
  }
  if (name === "kalshi_redeem_all_positions") {
    const walletService = getSolanaWalletService();
    const marketService2 = getDFlowMarketService();
    const tradeService = getDFlowTradeService();
    const walletInfo = await walletService.getWalletInfo(userId);
    if (!walletInfo) {
      throw new Error("No wallet found");
    }
    if (walletInfo.type === "imported" && !args.password) {
      throw new Error("Password required for imported wallet transactions");
    }
    const tokens = await walletService.getAllTokenAccounts(walletInfo.publicKey);
    const potentialPositions = tokens.filter(
      (t) => Number(t.balance) > 0 && t.mint !== config.tokens.USDC && t.mint !== config.tokens.SOL && t.mint !== config.tokens.USDT
    );
    const results = [];
    for (const position of potentialPositions) {
      try {
        const market = await marketService2.getMarketByMint(position.mint);
        if (market) {
          const redemptionStatus = await marketService2.checkRedemptionStatus(market.ticker);
          if (redemptionStatus.canRedeem) {
            const result = await tradeService.redeemWinnings(
              userId,
              position.mint,
              Number(position.balance),
              args.password
            );
            results.push({
              market: market.ticker,
              mint: position.mint,
              amount: Number(position.balance) / Math.pow(10, position.decimals),
              txSignature: result.txSignature,
              success: true
            });
          }
        }
      } catch (e) {
        results.push({
          mint: position.mint,
          success: false,
          error: e.message
        });
      }
    }
    const successCount = results.filter((r) => r.success).length;
    return {
      results,
      summary: {
        total: results.length,
        successful: successCount,
        failed: results.length - successCount
      },
      message: successCount > 0 ? `Successfully redeemed ${successCount} positions.` : "No positions were redeemed."
    };
  }
  throw new Error(`Unknown tool: ${name}`);
}

// src/mcp/http-handler.ts
var MCPHttpHandler = class {
  /**
   * Handle MCP requests
   */
  async handleRequest(req, res) {
    const body = req.body;
    if (!body || body.jsonrpc !== "2.0" || !body.method) {
      res.status(400).json({
        jsonrpc: "2.0",
        id: body?.id || null,
        error: {
          code: -32600,
          message: "Invalid JSON-RPC request"
        }
      });
      return;
    }
    try {
      switch (body.method) {
        case "initialize":
          res.json({
            jsonrpc: "2.0",
            id: body.id,
            result: {
              protocolVersion: "2024-11-05",
              serverInfo: {
                name: "quantish-kalshi-sdk",
                version: "1.0.0"
              },
              capabilities: {
                tools: {}
              }
            }
          });
          return;
        case "tools/list":
          res.json({
            jsonrpc: "2.0",
            id: body.id,
            result: {
              tools: kalshiTools
            }
          });
          return;
        case "tools/call":
          await this.handleToolCall(req, res, body);
          return;
        default:
          res.json({
            jsonrpc: "2.0",
            id: body.id,
            error: {
              code: -32601,
              message: `Method not found: ${body.method}`
            }
          });
          return;
      }
    } catch (error) {
      console.error("MCP Error:", error);
      res.json({
        jsonrpc: "2.0",
        id: body.id,
        error: {
          code: -32603,
          message: error.message || "Internal error"
        }
      });
    }
  }
  /**
   * Handle tool calls with authentication
   */
  async handleToolCall(req, res, body) {
    const { name, arguments: args } = body.params || {};
    if (!name) {
      res.json({
        jsonrpc: "2.0",
        id: body.id,
        error: {
          code: -32602,
          message: "Tool name required"
        }
      });
      return;
    }
    const tool = kalshiTools.find((t) => t.name === name);
    if (!tool) {
      res.json({
        jsonrpc: "2.0",
        id: body.id,
        error: {
          code: -32602,
          message: `Unknown tool: ${name}`
        }
      });
      return;
    }
    const context = {};
    const unauthenticatedTools = [
      "kalshi_signup",
      // Self-service signup - creates user + wallet + API key
      "kalshi_request_api_key",
      "kalshi_get_wallet_import_instructions",
      "kalshi_import_private_key",
      // Creates new user, no existing auth needed
      // Read-only market discovery (public data from DFlow)
      "kalshi_search_markets",
      "kalshi_get_market",
      "kalshi_get_events",
      "kalshi_get_live_data",
      "kalshi_check_market_initialization",
      "kalshi_check_redemption_status",
      "kalshi_get_market_by_mint"
    ];
    console.log(`[MCP] Tool call: ${name}, is unauthenticated: ${unauthenticatedTools.includes(name)}`);
    if (!unauthenticatedTools.includes(name)) {
      const apiKey = req.headers["x-api-key"];
      if (!apiKey) {
        res.json({
          jsonrpc: "2.0",
          id: body.id,
          error: {
            code: -32e3,
            message: "API key required"
          }
        });
        return;
      }
      const apiKeyService = getApiKeyService();
      const validation = await apiKeyService.validateApiKey(apiKey);
      if (!validation.isValid) {
        res.json({
          jsonrpc: "2.0",
          id: body.id,
          error: {
            code: -32e3,
            message: validation.message || "Invalid API key"
          }
        });
        return;
      }
      context.userId = validation.userId;
      const hmacSignature = req.headers["x-hmac-signature"];
      const hmacTimestamp = req.headers["x-hmac-timestamp"];
      if (hmacSignature && hmacTimestamp && validation.keyRecord?.apiSecret) {
        const bodyString = JSON.stringify(req.body);
        const isValidHmac = apiKeyService.validateHmacSignature(
          validation.keyRecord.apiSecret,
          hmacTimestamp,
          req.method,
          req.originalUrl,
          bodyString,
          hmacSignature
        );
        if (!isValidHmac) {
          res.json({
            jsonrpc: "2.0",
            id: body.id,
            error: {
              code: -32e3,
              message: "Invalid HMAC signature"
            }
          });
          return;
        }
      }
    }
    try {
      const result = await executeTool(name, args || {}, context);
      res.json({
        jsonrpc: "2.0",
        id: body.id,
        result: {
          content: [
            {
              type: "text",
              text: JSON.stringify(result, null, 2)
            }
          ]
        }
      });
    } catch (error) {
      console.error(`Tool ${name} error:`, error);
      res.json({
        jsonrpc: "2.0",
        id: body.id,
        error: {
          code: -32603,
          message: error.message || "Tool execution failed"
        }
      });
    }
  }
};
var handlerInstance = null;
function getMCPHttpHandler() {
  if (!handlerInstance) {
    handlerInstance = new MCPHttpHandler();
  }
  return handlerInstance;
}

// src/routes/mcp.ts
var router = Router();
var mcpHandler = getMCPHttpHandler();
router.post("/", async (req, res) => {
  await mcpHandler.handleRequest(req, res);
});
router.get("/info", (req, res) => {
  res.json({
    name: "Quantish Kalshi SDK",
    version: "1.0.0",
    description: "Kalshi prediction market trading via DFlow on Solana",
    protocol: "MCP 2024-11-05",
    endpoints: {
      mcp: "/mcp",
      info: "/mcp/info"
    },
    authentication: {
      required: true,
      method: "API Key in x-api-key header",
      hmac: "Optional HMAC signing supported"
    },
    features: [
      "Solana wallet generation",
      "Bring Your Own Wallet (encrypted import)",
      "Kalshi market discovery",
      "Prediction market trading via DFlow",
      "Position tracking",
      "Outcome redemption"
    ]
  });
});
var mcp_default = router;

// src/routes/admin.ts
import { Router as Router2 } from "express";
var router2 = Router2();
var adminAuth = (req, res, next) => {
  const adminKey = req.headers["x-admin-key"];
  if (!config.admin.apiKey) {
    return res.status(500).json({ error: "Admin API key not configured" });
  }
  if (!adminKey || adminKey !== config.admin.apiKey) {
    return res.status(401).json({ error: "Invalid admin API key" });
  }
  next();
};
router2.use(adminAuth);
router2.post("/access-codes", async (req, res) => {
  try {
    const accessCodeService = getAccessCodeService();
    const code = await accessCodeService.createAccessCode({
      ...req.body,
      createdBy: "admin"
    });
    res.json({ success: true, accessCode: code });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});
router2.get("/access-codes", async (req, res) => {
  try {
    const accessCodeService = getAccessCodeService();
    const codes = await accessCodeService.listAccessCodes();
    res.json({ accessCodes: codes });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});
router2.delete("/access-codes/:codeOrId", async (req, res) => {
  try {
    const accessCodeService = getAccessCodeService();
    const success = await accessCodeService.revokeAccessCode(req.params.codeOrId);
    res.json({ success });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});
router2.post("/access-codes/bulk", async (req, res) => {
  try {
    const { count = 10, ...options } = req.body;
    const accessCodeService = getAccessCodeService();
    const codes = [];
    for (let i = 0; i < count; i++) {
      const code = await accessCodeService.createAccessCode({
        ...options,
        createdBy: "admin"
      });
      codes.push(code);
    }
    res.json({ success: true, count: codes.length, accessCodes: codes });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});
var admin_default = router2;

// src/routes/bot.ts
import { Router as Router3 } from "express";
var router3 = Router3();
router3.post("/setup", async (req, res, next) => {
  try {
    const { externalId } = req.body;
    const prisma2 = getPrismaClient();
    if (!externalId) {
      return res.status(400).json({
        success: false,
        error: "externalId is required (your Privy ID like did:privy:xxx)"
      });
    }
    const walletService = getSolanaWalletService();
    const apiKeyService = getApiKeyService();
    let user = await prisma2.user.findUnique({
      where: { externalId }
    });
    let status = "EXISTING";
    let isNewUser = false;
    if (!user) {
      user = await prisma2.user.create({
        data: { externalId }
      });
      status = "CREATED";
      isNewUser = true;
      console.log(`New user created for ${externalId}: ${user.id}`);
    }
    let walletInfo = await walletService.getWalletInfo(user.id);
    let isNewWallet = false;
    if (!walletInfo) {
      walletInfo = await walletService.generateWallet(user.id);
      isNewWallet = true;
      console.log(`New wallet generated for ${externalId}: ${walletInfo.publicKey}`);
    }
    if (!isNewUser && !isNewWallet) {
      status = "READY";
    }
    const keyResult = await apiKeyService.createApiKey(user.id, "Mobile App");
    await prisma2.activityLog.create({
      data: {
        userId: user.id,
        action: isNewUser ? "USER_CREATED" : isNewWallet ? "WALLET_CREATED" : "CREDENTIALS_ISSUED",
        resource: "setup",
        details: {
          externalId,
          walletPublicKey: walletInfo.publicKey,
          keyPrefix: keyResult.keyPrefix,
          isNewUser,
          isNewWallet
        }
      }
    });
    res.json({
      success: true,
      userId: user.id,
      externalId: user.externalId,
      wallet: {
        publicKey: walletInfo.publicKey,
        type: walletInfo.type,
        network: "solana-mainnet"
      },
      mcpApiKey: keyResult.apiKey,
      mcpApiSecret: keyResult.apiSecret,
      status,
      message: status === "CREATED" ? "New account and wallet created. Fund your wallet with SOL and USDC to start trading." : status === "READY" ? "Account ready. New API key issued." : "Wallet created for existing account.",
      nextSteps: {
        fundWallet: `Send SOL (for gas) and USDC (for trading) to ${walletInfo.publicKey}`,
        mcpEndpoint: "https://kalshi-mcp-server-production.up.railway.app/mcp",
        documentation: "Use x-api-key header with your mcpApiKey for authenticated requests"
      }
    });
  } catch (error) {
    console.error("Setup error:", error);
    next(error);
  }
});
router3.post("/credentials", async (req, res, next) => {
  try {
    const { externalId } = req.body;
    const prisma2 = getPrismaClient();
    if (!externalId) {
      return res.status(400).json({
        success: false,
        error: "externalId is required"
      });
    }
    const user = await prisma2.user.findUnique({
      where: { externalId },
      include: {
        apiKeys: {
          where: { isActive: true },
          orderBy: { createdAt: "desc" },
          take: 1
        }
      }
    });
    if (!user) {
      return res.json({
        hasWallet: false,
        walletPublicKey: null,
        mcpApiKey: null,
        message: "User not found. Call kalshi_request_api_key first to create an account."
      });
    }
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(user.id);
    const apiKeyService = getApiKeyService();
    const keyResult = await apiKeyService.createApiKey(user.id, "Mobile App");
    await prisma2.activityLog.create({
      data: {
        userId: user.id,
        action: "BOT_CREDENTIALS_ISSUED",
        resource: "api_key",
        resourceId: keyResult.keyId,
        details: {
          keyPrefix: keyResult.keyPrefix,
          source: "mobile_app"
        }
      }
    });
    res.json({
      hasWallet: !!walletInfo,
      walletType: walletInfo?.type || null,
      walletPublicKey: walletInfo?.publicKey || null,
      mcpApiKey: keyResult.apiKey,
      mcpApiSecret: keyResult.apiSecret,
      message: walletInfo ? "Credentials issued successfully. Save the API key - it cannot be recovered." : "User exists but wallet not set up. Call kalshi_setup_wallet to create a wallet."
    });
  } catch (error) {
    next(error);
  }
});
router3.get("/credentials/:externalId", async (req, res, next) => {
  try {
    const { externalId } = req.params;
    const prisma2 = getPrismaClient();
    const user = await prisma2.user.findUnique({
      where: { externalId },
      include: {
        apiKeys: {
          where: { isActive: true },
          select: {
            id: true,
            keyPrefix: true,
            name: true,
            createdAt: true
          }
        }
      }
    });
    if (!user) {
      return res.json({
        exists: false,
        hasWallet: false
      });
    }
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(user.id);
    res.json({
      exists: true,
      hasWallet: !!walletInfo,
      walletType: walletInfo?.type || null,
      walletPublicKey: walletInfo?.publicKey || null,
      activeKeyCount: user.apiKeys.length,
      apiKeyPrefixes: user.apiKeys.map((k) => k.keyPrefix)
    });
  } catch (error) {
    next(error);
  }
});
router3.get("/balances/:externalId", async (req, res, next) => {
  try {
    const { externalId } = req.params;
    const prisma2 = getPrismaClient();
    const user = await prisma2.user.findUnique({
      where: { externalId }
    });
    if (!user) {
      return res.status(404).json({
        success: false,
        error: "User not found"
      });
    }
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(user.id);
    if (!walletInfo) {
      return res.status(404).json({
        success: false,
        error: "No wallet found for user"
      });
    }
    const balances = await walletService.getBalances(walletInfo.publicKey);
    res.json({
      success: true,
      publicKey: walletInfo.publicKey,
      walletType: walletInfo.type,
      balances
    });
  } catch (error) {
    next(error);
  }
});
router3.get("/status/:externalId", async (req, res, next) => {
  try {
    const { externalId } = req.params;
    const prisma2 = getPrismaClient();
    const user = await prisma2.user.findUnique({
      where: { externalId },
      include: {
        apiKeys: {
          where: { isActive: true },
          select: { id: true }
        }
      }
    });
    if (!user) {
      return res.json({
        externalId,
        walletAddress: null,
        hasWallet: false,
        hasApiKey: false,
        status: "NOT_FOUND",
        isReady: false,
        message: "User not found. Call POST /api/bot/setup first."
      });
    }
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(user.id);
    if (!walletInfo) {
      return res.json({
        externalId,
        walletAddress: null,
        hasWallet: false,
        hasApiKey: user.apiKeys.length > 0,
        status: "NO_WALLET",
        isReady: false,
        message: "User exists but no wallet. Call POST /api/bot/setup to create one."
      });
    }
    const balances = await walletService.getBalances(walletInfo.publicKey);
    const hasSol = balances.sol > 1e-3;
    const hasUsdc = balances.usdc > 0;
    let status;
    let isReady;
    if (!hasSol && !hasUsdc) {
      status = "NEEDS_FUNDING";
      isReady = false;
    } else if (!hasSol) {
      status = "NEEDS_SOL";
      isReady = false;
    } else if (!hasUsdc) {
      status = "NEEDS_USDC";
      isReady = true;
    } else {
      status = "READY";
      isReady = true;
    }
    res.json({
      externalId,
      walletAddress: walletInfo.publicKey,
      walletType: walletInfo.type,
      hasWallet: true,
      hasApiKey: user.apiKeys.length > 0,
      balances: {
        sol: balances.sol,
        usdc: balances.usdc
      },
      status,
      isReady,
      message: isReady ? "Wallet ready for trading" : `Wallet needs funding. Send assets to ${walletInfo.publicKey}`
    });
  } catch (error) {
    next(error);
  }
});
var bot_default = router3;

// src/routes/users.ts
import { Router as Router4 } from "express";
var router4 = Router4();
router4.get("/:userId", async (req, res, next) => {
  try {
    const { userId } = req.params;
    const prisma2 = getPrismaClient();
    const user = await prisma2.user.findUnique({
      where: { id: userId },
      select: {
        id: true,
        externalId: true,
        solanaPublicKey: true,
        importedWalletPublicKey: true,
        walletImportedAt: true,
        createdAt: true,
        updatedAt: true
      }
    });
    if (!user) {
      return res.status(404).json({
        success: false,
        error: "User not found"
      });
    }
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(userId);
    res.json({
      success: true,
      data: {
        userId: user.id,
        externalId: user.externalId,
        wallet: walletInfo,
        createdAt: user.createdAt
      }
    });
  } catch (error) {
    next(error);
  }
});
router4.get("/external/:externalId", async (req, res, next) => {
  try {
    const { externalId } = req.params;
    const prisma2 = getPrismaClient();
    const user = await prisma2.user.findUnique({
      where: { externalId },
      select: {
        id: true,
        externalId: true,
        solanaPublicKey: true,
        importedWalletPublicKey: true,
        createdAt: true
      }
    });
    if (!user) {
      return res.status(404).json({
        success: false,
        error: "User not found"
      });
    }
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(user.id);
    res.json({
      success: true,
      data: {
        userId: user.id,
        externalId: user.externalId,
        wallet: walletInfo,
        createdAt: user.createdAt
      }
    });
  } catch (error) {
    next(error);
  }
});
router4.post("/:userId/setup-wallet", async (req, res, next) => {
  try {
    const { userId } = req.params;
    const walletService = getSolanaWalletService();
    const existingWallet = await walletService.getWalletInfo(userId);
    if (existingWallet) {
      return res.json({
        success: true,
        data: {
          message: "Wallet already exists",
          wallet: existingWallet
        }
      });
    }
    const wallet = await walletService.generateWallet(userId);
    res.json({
      success: true,
      data: {
        message: "Solana wallet generated successfully",
        wallet,
        note: "Fund this address with SOL and USDC to start trading."
      }
    });
  } catch (error) {
    next(error);
  }
});
router4.get("/:userId/balances", async (req, res, next) => {
  try {
    const { userId } = req.params;
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(userId);
    if (!walletInfo) {
      return res.status(404).json({
        success: false,
        error: "No wallet found"
      });
    }
    const balances = await walletService.getBalances(walletInfo.publicKey);
    res.json({
      success: true,
      data: {
        publicKey: walletInfo.publicKey,
        walletType: walletInfo.type,
        balances
      }
    });
  } catch (error) {
    next(error);
  }
});
router4.get("/:userId/tokens", async (req, res, next) => {
  try {
    const { userId } = req.params;
    const walletService = getSolanaWalletService();
    const walletInfo = await walletService.getWalletInfo(userId);
    if (!walletInfo) {
      return res.status(404).json({
        success: false,
        error: "No wallet found"
      });
    }
    const tokens = await walletService.getAllTokenAccounts(walletInfo.publicKey);
    res.json({
      success: true,
      data: {
        publicKey: walletInfo.publicKey,
        tokens
      }
    });
  } catch (error) {
    next(error);
  }
});
var users_default = router4;

// src/routes/orders.ts
import { Router as Router5 } from "express";
var router5 = Router5();
router5.get("/", async (req, res, next) => {
  try {
    const userId = req.userId;
    const { status, limit = "50", offset = "0" } = req.query;
    const prisma2 = getPrismaClient();
    if (!userId) {
      return res.status(401).json({
        success: false,
        error: "Authentication required"
      });
    }
    const orders = await prisma2.order.findMany({
      where: {
        userId,
        ...status ? { status } : {}
      },
      orderBy: { createdAt: "desc" },
      take: parseInt(limit),
      skip: parseInt(offset)
    });
    const total = await prisma2.order.count({
      where: {
        userId,
        ...status ? { status } : {}
      }
    });
    res.json({
      success: true,
      data: {
        orders,
        pagination: {
          total,
          limit: parseInt(limit),
          offset: parseInt(offset)
        }
      }
    });
  } catch (error) {
    next(error);
  }
});
router5.get("/:orderId", async (req, res, next) => {
  try {
    const userId = req.userId;
    const { orderId } = req.params;
    const prisma2 = getPrismaClient();
    if (!userId) {
      return res.status(401).json({
        success: false,
        error: "Authentication required"
      });
    }
    const order = await prisma2.order.findFirst({
      where: {
        id: orderId,
        userId
      }
    });
    if (!order) {
      return res.status(404).json({
        success: false,
        error: "Order not found"
      });
    }
    res.json({
      success: true,
      data: order
    });
  } catch (error) {
    next(error);
  }
});
var orders_default = router5;

// src/routes/positions.ts
import { Router as Router6 } from "express";
var router6 = Router6();
router6.get("/", async (req, res, next) => {
  try {
    const userId = req.userId;
    const { settled } = req.query;
    const prisma2 = getPrismaClient();
    if (!userId) {
      return res.status(401).json({
        success: false,
        error: "Authentication required"
      });
    }
    const positions = await prisma2.position.findMany({
      where: {
        userId,
        ...settled !== void 0 ? { isSettled: settled === "true" } : {}
      },
      orderBy: { updatedAt: "desc" }
    });
    res.json({
      success: true,
      data: { positions }
    });
  } catch (error) {
    next(error);
  }
});
router6.get("/:positionId", async (req, res, next) => {
  try {
    const userId = req.userId;
    const { positionId } = req.params;
    const prisma2 = getPrismaClient();
    if (!userId) {
      return res.status(401).json({
        success: false,
        error: "Authentication required"
      });
    }
    const position = await prisma2.position.findFirst({
      where: {
        id: positionId,
        userId
      }
    });
    if (!position) {
      return res.status(404).json({
        success: false,
        error: "Position not found"
      });
    }
    res.json({
      success: true,
      data: position
    });
  } catch (error) {
    next(error);
  }
});
router6.get("/market/:marketTicker", async (req, res, next) => {
  try {
    const userId = req.userId;
    const { marketTicker } = req.params;
    const prisma2 = getPrismaClient();
    if (!userId) {
      return res.status(401).json({
        success: false,
        error: "Authentication required"
      });
    }
    const positions = await prisma2.position.findMany({
      where: {
        userId,
        marketTicker
      }
    });
    res.json({
      success: true,
      data: { positions }
    });
  } catch (error) {
    next(error);
  }
});
var positions_default = router6;

// src/routes/markets.ts
import { Router as Router7 } from "express";
var router7 = Router7();
var marketService = getDFlowMarketService();
router7.get("/", async (req, res, next) => {
  try {
    const { status, eventTicker, limit = "20", cursor } = req.query;
    const result = await marketService.getMarkets({
      status,
      eventTicker,
      limit: parseInt(limit),
      cursor
    });
    res.json({
      success: true,
      data: result
    });
  } catch (error) {
    next(error);
  }
});
router7.get("/search", async (req, res, next) => {
  try {
    const { q, limit = "10", status } = req.query;
    if (!q) {
      return res.status(400).json({
        success: false,
        error: 'Query parameter "q" is required'
      });
    }
    const events = await marketService.searchEvents(
      q,
      parseInt(limit),
      status
    );
    res.json({
      success: true,
      query: q,
      data: { events }
    });
  } catch (error) {
    next(error);
  }
});
router7.get("/events", async (req, res, next) => {
  try {
    const { category, marketStatus, limit = "10", cursor } = req.query;
    const result = await marketService.getEvents({
      category,
      marketStatus,
      limit: parseInt(limit),
      cursor,
      withNestedMarkets: true
    });
    res.json({
      success: true,
      data: result
    });
  } catch (error) {
    next(error);
  }
});
router7.get("/categories", async (req, res, next) => {
  try {
    const categories = await marketService.getTagsByCategories();
    res.json({
      success: true,
      data: categories
    });
  } catch (error) {
    console.error("Categories endpoint error:", error.message);
    res.json({
      success: true,
      data: {
        note: "Categories endpoint not fully supported by DFlow API",
        suggestedCategories: [
          "Politics",
          "Sports",
          "Economics",
          "Finance",
          "Weather",
          "Entertainment",
          "Technology",
          "Science"
        ]
      }
    });
  }
});
router7.get("/:ticker", async (req, res, next) => {
  try {
    const { ticker } = req.params;
    const market = await marketService.getMarket(ticker);
    if (!market) {
      return res.status(404).json({
        success: false,
        error: "Market not found"
      });
    }
    res.json({
      success: true,
      data: market
    });
  } catch (error) {
    next(error);
  }
});
router7.get("/:ticker/live", async (req, res, next) => {
  try {
    const { ticker } = req.params;
    const liveData = await marketService.getLiveData(ticker);
    if (!liveData) {
      return res.status(404).json({
        success: false,
        error: "Market not found"
      });
    }
    res.json({
      success: true,
      data: liveData
    });
  } catch (error) {
    next(error);
  }
});
router7.get("/:ticker/trades", async (req, res, next) => {
  try {
    const { ticker } = req.params;
    const { limit = "20", cursor } = req.query;
    const trades = await marketService.getTrades(ticker, {
      limit: parseInt(limit),
      cursor
    });
    res.json({
      success: true,
      data: { trades }
    });
  } catch (error) {
    next(error);
  }
});
router7.get("/:ticker/candlesticks", async (req, res, next) => {
  try {
    const { ticker } = req.params;
    const { resolution = "1h", from, to } = req.query;
    const candlesticks = await marketService.getMarketCandlesticks(ticker, {
      resolution,
      from,
      to
    });
    res.json({
      success: true,
      data: { candlesticks }
    });
  } catch (error) {
    next(error);
  }
});
router7.get("/events/:ticker", async (req, res, next) => {
  try {
    const { ticker } = req.params;
    const event = await marketService.getEvent(ticker);
    res.json({
      success: true,
      data: event
    });
  } catch (error) {
    next(error);
  }
});
var markets_default = router7;

// src/middleware/auth.ts
async function apiKeyAuth(req, res, next) {
  try {
    const apiKey = req.headers["x-api-key"];
    if (!apiKey) {
      res.status(401).json({
        success: false,
        error: "API key required. Provide x-api-key header."
      });
      return;
    }
    const apiKeyService = getApiKeyService();
    const validation = await apiKeyService.validateApiKey(apiKey);
    if (!validation.isValid) {
      res.status(401).json({
        success: false,
        error: validation.message || "Invalid API key"
      });
      return;
    }
    req.userId = validation.userId;
    req.keyId = validation.keyRecord?.id;
    next();
  } catch (error) {
    console.error("Auth middleware error:", error);
    res.status(500).json({
      success: false,
      error: "Authentication error"
    });
  }
}

// src/index.ts
validateConfig();
var app = express();
app.use(helmet());
app.use(cors({
  origin: "*",
  // Configure based on your needs
  methods: ["GET", "POST", "DELETE", "OPTIONS"],
  allowedHeaders: ["Content-Type", "x-api-key", "x-admin-key", "x-hmac-signature", "x-hmac-timestamp"]
}));
var limiter = rateLimit({
  windowMs: config.rateLimit.windowMs,
  max: config.rateLimit.maxRequests,
  message: { error: "Rate limit exceeded. Please slow down." },
  standardHeaders: true,
  legacyHeaders: false,
  keyGenerator: (req) => {
    return req.headers["x-api-key"] || req.ip || "unknown";
  }
});
var registrationLimiter = rateLimit({
  windowMs: 60 * 60 * 1e3,
  // 1 hour
  max: 5,
  message: { error: "Too many registration attempts. Try again later." },
  standardHeaders: true,
  legacyHeaders: false
});
app.use(express.json({ limit: "1mb", strict: true }));
app.use(express.urlencoded({ extended: true, limit: "1mb" }));
app.get("/health", (req, res) => {
  res.json({
    status: "healthy",
    service: "quantish-kalshi-sdk",
    version: "1.0.0",
    timestamp: (/* @__PURE__ */ new Date()).toISOString()
  });
});
app.get("/", (req, res) => {
  res.json({
    name: "Quantish Kalshi SDK",
    version: "1.0.0",
    description: "Kalshi prediction market trading via DFlow on Solana",
    endpoints: {
      health: "/health",
      mcp: "/mcp",
      mcpInfo: "/mcp/info",
      bot: "/api/bot (mobile app credentials)",
      markets: "/api/markets (public)",
      users: "/api/users (requires x-api-key)",
      orders: "/api/orders (requires x-api-key)",
      positions: "/api/positions (requires x-api-key)",
      admin: "/api/admin (requires x-admin-key)"
    },
    documentation: "https://github.com/quantish/kalshi-sdk",
    features: [
      "Solana wallet generation",
      "Bring Your Own Wallet (encrypted import from Phantom/Solflare)",
      "Kalshi market discovery via DFlow",
      "Prediction market trading",
      "Position tracking",
      "Outcome redemption",
      "REST API & MCP interface",
      "Mobile app integration"
    ]
  });
});
app.use("/mcp", limiter, mcp_default);
app.use("/api/admin", admin_default);
app.use("/api/bot", bot_default);
app.use("/api/markets", markets_default);
app.use("/api/users", apiKeyAuth, users_default);
app.use("/api/orders", apiKeyAuth, orders_default);
app.use("/api/positions", apiKeyAuth, positions_default);
app.use((err, req, res, next) => {
  console.error("Server error:", err);
  res.status(500).json({
    error: "Internal server error",
    message: config.server.nodeEnv === "development" ? err.message : void 0
  });
});
app.use((req, res) => {
  res.status(404).json({
    error: "Not found",
    path: req.path
  });
});
var PORT = config.server.port;
var server = app.listen(PORT, () => {
  console.log(`
\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557
\u2551                                                               \u2551
\u2551   \u{1F680} Quantish Kalshi SDK Server                               \u2551
\u2551                                                               \u2551
\u2551   Version:     1.0.3 (DB Retry Fix)                               \u2551
\u2551   Environment: ${config.server.nodeEnv.padEnd(20)}             \u2551
\u2551   Port:        ${PORT.toString().padEnd(20)}                   \u2551
\u2551                                                               \u2551
\u2551   Endpoints:                                                  \u2551
\u2551   \u2022 MCP:    http://localhost:${PORT}/mcp                       \u2551
\u2551   \u2022 Info:   http://localhost:${PORT}/mcp/info                  \u2551
\u2551   \u2022 Health: http://localhost:${PORT}/health                    \u2551
\u2551   \u2022 Admin:  http://localhost:${PORT}/api/admin                 \u2551
\u2551                                                               \u2551
\u2551   Features:                                                   \u2551
\u2551   \u2022 Kalshi market trading via DFlow API                       \u2551
\u2551   \u2022 Solana wallet generation                                  \u2551
\u2551   \u2022 Bring Your Own Wallet (encrypted import)                  \u2551
\u2551   \u2022 Position tracking & redemption                            \u2551
\u2551                                                               \u2551
\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D
  `);
});
var shutdown = async () => {
  console.log("\nShutting down gracefully...");
  server.close(() => {
    console.log("HTTP server closed");
  });
  await disconnectPrisma();
  console.log("Database connection closed");
  process.exit(0);
};
process.on("SIGTERM", shutdown);
process.on("SIGINT", shutdown);
export {
  app
};
